{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "3b7a5883",
   "metadata": {},
   "outputs": [],
   "source": [
    "import json\n",
    "import matplotlib.pyplot as plt\n",
    "import numpy as np\n",
    "from matplotlib import rc\n",
    "import os\n",
    "rc('font',**{'family':'sans-serif','sans-serif':'Arial'})\n",
    "rc('text',usetex=True)\n",
    "letters = ['A','T','C','G']\n",
    "filelabels = ['CDS','UTR']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c4033ef9",
   "metadata": {},
   "outputs": [],
   "source": [
    "files = os.listdir('./ProteinCodingOrthologs/downloaded')\n",
    "path = './ProteinCodingOrthologs/downloaded'\n",
    "Archetypes = ['GCF_015852505.1_mTacAcu1.pri', 'GCF_016433145.1_AgileGrace', 'GCA_016432865.2_AdamAnt_v2',\n",
    "         'GCF_000313985.2_ASM31398v2', 'GCA_000152225.2_Pcap_2.0', 'GCF_014898055.1_MPIMG_talOcc4',\n",
    "         'GCF_000296755.1_EriEur2.0', 'GCA_029834395.2_SorCin_2.1', 'Human_NEAT1', 'GCA_013371645.1_OryCun3.0',\n",
    "         'GCF_014633375.1_OchPri4.0', 'GCF_020740685.1_mJacJac1.mat.Y.cur', 'GCA_023101885.1_KIZ_TCIN_v1.0',\n",
    "         'Mouse_NEAT1','GCF_019054845.1_ASM1905484v1', 'GCA_004027875.1_AplRuf_v1_BIUU']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f30d416d",
   "metadata": {},
   "outputs": [],
   "source": [
    "latin = {}\n",
    "genera = {}\n",
    "species = {}\n",
    "arch = {}\n",
    "with open('Metadata') as file1:\n",
    "    for line in file1:\n",
    "        dat = line.strip().split('\\t')\n",
    "        if not line.startswith('#'):\n",
    "            if dat[3] == '1' or dat[4] == '1':\n",
    "                latin[dat[0]] = dat[20]\n",
    "                species[dat[20]] = ''\n",
    "                genus = dat[20].split(' ')[0]\n",
    "                if genus not in genera:\n",
    "                    genera[genus] = set()\n",
    "                genera[genus].add(dat[20].split(' ')[1])\n",
    "                if dat[0] in Archetypes:\n",
    "                    arch[dat[20]] = dat[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fde272c2",
   "metadata": {},
   "outputs": [],
   "source": [
    "filenames = ['cds','3p_utr']\n",
    "filelabels = ['CDS','UTR']\n",
    "#Going through downloaded fasta files with protein-coding genes orthologs one by one\n",
    "for gene in files:\n",
    "    #Files are separate for CDS and 3p-UTR\n",
    "    for j in range(2):\n",
    "        filename = filenames[j]\n",
    "        filelabel = filelabels[j]\n",
    "        #temporary dict for found species and genera in orthologs fasta\n",
    "        a = {}\n",
    "        ag = {}\n",
    "        with open(path + '/' + gene + '/' + filename + '.fna') as file1:\n",
    "            for line in file1:\n",
    "                if line.startswith('>'):\n",
    "                    dat = line.strip().split('[')\n",
    "                    flag = 0\n",
    "                    for item in dat:\n",
    "                        if 'transcript' in item:\n",
    "                            if 'X1' not in item:\n",
    "                                flag = 1\n",
    "                    if flag == 0:\n",
    "                        for item in dat:\n",
    "                            if 'organism' in item:\n",
    "                                name = ' '.join(item.replace(']','').split('=')[1].split(' ')[:2])\n",
    "                                name_genus = name.split(' ')[0]\n",
    "                                a[name] = line\n",
    "                                if name_genus not in ag:\n",
    "                                    ag[name_genus] = line\n",
    "        #Select only species and genera which are part of our dataset with NEAT1 orthologs\n",
    "        #Separately for all NEAT1 orthologs and archetypes\n",
    "        sel = {}\n",
    "        selarch = {}\n",
    "        for item in arch:\n",
    "            if item in a:\n",
    "                selarch[a[item]] = ''\n",
    "            else:\n",
    "                if item.split(' ')[0] in ag:\n",
    "                    selarch[ag[item.split(' ')[0]]] = ''\n",
    "        for item in species:\n",
    "            if item in a:\n",
    "                sel[a[item]] = ''\n",
    "            else:\n",
    "                if item.split(' ')[0] in ag:\n",
    "                    sel[ag[item.split(' ')[0]]] = ''\n",
    "        \n",
    "        #Leave only genes with more than 150 genera and sequences of genera relevant for our study only \n",
    "        if len(sel) >= 150:\n",
    "            out = open('./Orthologs/all/' + filelabel + '/' + gene + '_' + filelabel + '.fasta','w')                        \n",
    "            flag = 0\n",
    "            with open(path + '/' + gene + '/' + filename + '.fna') as file1:\n",
    "                for line in file1:\n",
    "                    if line.startswith('>') and line in sel:\n",
    "                        dat = line.strip().split('[')\n",
    "                        for item in dat:\n",
    "                            if 'organism' in item:\n",
    "                                name = ' '.join(item.replace(']','').split('=')[1].split(' ')[:2])\n",
    "                        out.write(line.split(' ')[0] + ' ' + gene + ':' + name + '\\n')\n",
    "                        flag = 1\n",
    "                    elif line.startswith('>') and line not in sel:\n",
    "                        flag = 0\n",
    "                    elif not line.startswith('>') and flag == 1:\n",
    "                        out.write(line)  \n",
    "            out.close()\n",
    "        #For archtypes we applied threshold of at least 10 orthologs\n",
    "        if len(selarch) >= 10:\n",
    "            out = open('./Orthologs/archetypes/' + filelabel + '/' + gene + '_' + filelabel + '.fasta','w')                        \n",
    "            flag = 0\n",
    "            with open(path + '/' + gene + '/' + filename + '.fna') as file1:\n",
    "                for line in file1:\n",
    "                    if line.startswith('>') and line in selarch:\n",
    "                        dat = line.strip().split('[')\n",
    "                        for item in dat:\n",
    "                            if 'organism' in item:\n",
    "                                name = ' '.join(item.replace(']','').split('=')[1].split(' ')[:2])\n",
    "                        out.write(line.split(' ')[0] + ' ' + gene + ':' + name + '\\n')\n",
    "                        flag = 1\n",
    "                    elif line.startswith('>') and line not in selarch:\n",
    "                        flag = 0\n",
    "                    elif not line.startswith('>') and flag == 1:\n",
    "                        out.write(line)  \n",
    "            out.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8d1ce074",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Analyse the nucleotide usage in orthologs of selected genes\n",
    "filelabels = ['CDS','UTR']\n",
    "d = {}\n",
    "\n",
    "for filelabel in filelabels:\n",
    "    d[filelabel] = {}\n",
    "    for gene in files:\n",
    "        try:\n",
    "            d[filelabel][gene] = {}\n",
    "            with open('./Orthologs/all/' + filelabel + '/' + gene + '_' + filelabel + '.fasta') as file1:\n",
    "                fl = 0\n",
    "                for line in file1:\n",
    "                    if line.startswith('>') and fl == 0:\n",
    "                        l = []\n",
    "                        name = line.rstrip().split(':')[-1]\n",
    "                        fl = 1\n",
    "                    elif line.startswith('>') and fl == 1:\n",
    "                        d[filelabel][gene][name] = {}\n",
    "                        seq = ''.join(l)\n",
    "                        length = len(seq)\n",
    "                        for letter in ['A','T','C','G']:\n",
    "                            d[filelabel][gene][name][letter] = round((seq.count(letter)/length)*100,2)\n",
    "                        l = []\n",
    "                        name = line.rstrip().split(':')[-1]\n",
    "                    elif not line.startswith('>'):\n",
    "                        l.append(line.strip())\n",
    "            d[filelabel][gene][name] = {}\n",
    "            seq = ''.join(l)\n",
    "            length = len(seq)\n",
    "            for letter in letters:\n",
    "                d[filelabel][gene][name][letter] = round((seq.count(letter)/length)*100,2)\n",
    "        except:\n",
    "            pass\n",
    "json.dump(d, open('./Orthologs/NucleotideUsage_CDS_UTR.json',\"w\"))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fa2086fe",
   "metadata": {},
   "source": [
    "### Make a plot from Fig.5E"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "aa4d2974",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "dict_keys(['CDS', 'UTR'])\n"
     ]
    }
   ],
   "source": [
    "d = json.load(open('./Orthologs/NucleotideUsage_CDS_UTR.json',\"r\"))\n",
    "print(d.keys())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "801f0fe5",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['A3GALT2', 'ABCC6', 'ABHD14A', 'ABRAXAS1', 'ACAD10', 'ACADS', 'ACO2', 'ACOT11', 'ACTC1', 'ACVR2B']\n",
      "{'A': 13.49, 'T': 17.01, 'C': 33.92, 'G': 35.58}\n"
     ]
    }
   ],
   "source": [
    "print(list(d['CDS'].keys())[:10])\n",
    "print(d['CDS']['A3GALT2']['Ursus americanus'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "51623f16",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Make a list with species, one per family\n",
    "families2 = []\n",
    "with open('./PhyloTree/leaves_order_one_per_family') as file1:\n",
    "    for line in file1:\n",
    "        dat = line.strip().split('\\t')\n",
    "        families2.append(dat[1])\n",
    "#Parse genes FASTA files\n",
    "seq = {}\n",
    "seq_file = './NEAT1_gene_all.fasta'\n",
    "with open(seq_file) as file1:\n",
    "    fl = 0\n",
    "    for line in file1:\n",
    "        if line.startswith('>') and fl == 0:\n",
    "            l = []\n",
    "            name = line.rstrip().lstrip('>')\n",
    "            fl = 1\n",
    "        elif line.startswith('>') and fl == 1:\n",
    "            seq[name] = ''.join(l)\n",
    "            l = []\n",
    "            name = line.rstrip().lstrip('>')\n",
    "        elif not line.startswith('>'):\n",
    "            l.append(line.strip())\n",
    "    seq[name] = ''.join(l)\n",
    "\n",
    "seq2 = {}\n",
    "seq_file = './MALAT1_gene_all.fasta'\n",
    "with open(seq_file) as file1:\n",
    "    fl = 0\n",
    "    for line in file1:\n",
    "        if line.startswith('>') and fl == 0:\n",
    "            l = []\n",
    "            name = line.rstrip().lstrip('>')\n",
    "            fl = 1\n",
    "        elif line.startswith('>') and fl == 1:\n",
    "            seq2[name] = ''.join(l)\n",
    "            l = []\n",
    "            name = line.rstrip().lstrip('>')\n",
    "        elif not line.startswith('>'):\n",
    "            l.append(line.strip())\n",
    "    seq2[name] = ''.join(l)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "9d4e2480",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Reformat dictionary d into more convenient for plotting form\n",
    "genelist = list(set(list(d['CDS'].keys()) + list(d['UTR'].keys())))\n",
    "val = {}\n",
    "for filelabel in filelabels:\n",
    "    val[filelabel] = {}\n",
    "    for gene in genelist:\n",
    "        if gene in d[filelabel]:\n",
    "            val[filelabel][gene] = {i:[] for i in letters}\n",
    "            for name in d[filelabel][gene]:\n",
    "                for letter in letters:\n",
    "                    val[filelabel][gene][letter].append(d[filelabel][gene][name][letter])\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "62900ad0",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAaAAAADQCAYAAABBY7UsAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjQuMywgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy/MnkTPAAAACXBIWXMAAAsTAAALEwEAmpwYAAAeVElEQVR4nO3dX2xc53nn8e8jynIUMxZFWZGMorFNyXu5oUl5r9o6tSkHyHY3hU3JQDbxRRFTKgIhziKV4sX6ygEUukLjrbGoSbm+cJwA+hMXAYIAjihX2AB7saZoNou9aVcj2cnGkm1SdCTX1h/q2Yv3PeThcDicQ86Zw5n5fQBC5Dlnznnpk+jR+77P+7zm7oiIiDTauqIbICIi7UkBSERECqEAJCIihVAAEhGRQigAiYhIIRSARESkEOsb8RAzG4jfdgEPAsfcfSKeGwbOAceBvcC0u59sRLtERKQ4jeoBnQDGY2B5Czhadn4EOA90KfiIiLSHhvSAgH53n4nfdwPTqXNvubs1qB0iIrJGNKQH5O6l1I97gOHya8ysrxFtERGRtaFhSQhm1mNmB4ET7j6WOtUT54hKZja8VCAysyEzG49fQw1ptIiI5MYaWQvOzLoIvZ9TleZ6zKwnntvRsEaJiEghGhqAYC4j7hSw2d1nzKwrNT+EmbnmhEREWl/uQ3BmNmBmZ1OHxuOf3TEYnc67DSIisvY0IgtuGjiW+nkXUHL3kplNA4eTE2Y2CCgNW0SkDTRkCC4Glu74Yz8wnGTGxaSDAWAG2OHuh3JvkIiIFK7hc0Ai7ebAgQNnAF588cUvFdsSqacDBw68APTW+baTL7744tN1vuea1aiFqCIirab39ttvf2jb9m11udmli5e4du1aTdea2Qjzo0oldz9kZpeBUULJs27gUGqkaQToAUoskYVcBAUgEZEV2rZ9G9948sm63OtHr77Ku++8u+x1ZnaKMI0xVnaqlExhxCUvp4H+OAVyzt33pc6tCaqG3SBvf+2eM29/7Z4zRbdD8qH3K40Q10pSIfgsEJe2jMRF+zPA7iTwpJe9FE0BSESkeSTDaLUoEepwjhEKPp8ws7NrqeyZhuBERJpHiRCEatFD2OqGOOdzMvagThCykQunHpCISJNIJRUMLHctsA8YTYbtoumlLi6CekAiIit06eIlfvTqq3W7Vy3cfbeZjZjZvngoST7oiRt8dhGy4J6K5c56YuLCDKFX9FRdGlwHCkAiIiszee3atZoy17Lcs5aLkoy2smObl7h2Ati9umblQwFIRGQF2mnBaF5UCSEnb3/tnhdYuEo6+X6yjo+ZfOAn7zy90g9rhf7KVHi3UOX9/mzTn/QCfPWj/7HoXBWrercizUBJCBk89vTxM489ffxMjZf3Uv8yHY28f9vJ8H570bsVWTUNweVr8oGfvPMlCAsVAZKfV0uLHrOrc49vMv0uq73fV+Jza333erfSLtQDEhGRQjSkB5TKWe8CHgSOxcyMpLTEIDAB9AGja6lUhNSf5p6kFaga9uo1agjuBHBfzEkHOMr8StwRd98NYGYlYJiwgKouHnv6+AvU738kvfGeZ8qOT77+wt6n6/QMqVHWd7uVzt74uTNLXFLpvN6tLKX3Mxs6Htq+5Y663Ozi1Md8en226jVJkVF3708dO+vu/bHuW5JuPQAk9eJG3H2sWrXsojQqAPWnejXdxNW4ZSt0ibuk7qWOAYj5Cd3JOt6z/P5SjF70bqVA27fcwV/82b+ty71e+fmvufDe71f8eXcfJQQYzOycu+8puyRdLXtNlORpSAAqi7J7CL0cCENuM+XXm1lPnSPz5Osv7P3Sam+S/Ms4fa8MWXGSj5rf7YEDvzoDC99fWvn71buVVhR7UYPM95AK07AkhFgO4iBwIlVKvLvCpdOELmIhDhw4cCaZoxARaSE9cWO60wBJb6hIDQtAsUczyvwGSZmY2ZCZjcevofq3cO3KuP5Imozer9QqTmV0rfDjpVjCZw8hGaxwDV0HFJMQTgCnzGwzlXs73VQYlkuPbzajeq3/kbVJ71caaMbMupJCo9S+PxAwN9c+bWYDy21sl7fcA1BMwR5OZW2Mxz+7CanXi4bhis7MEBGpxcWpj3nl57+u271qtAc4ambTxKrXK3jcIeKW3Sv4bN2sKACZ2cPMp/udcvc3q1w+DRxL/byL0BVM9rVI37cHOL6SNjVCPRIZZO3S+5WMJj+9PruqzLVK91zugvh3Z3mGW/k1Oyoc6099P8Ma2JQucwAys+8CjwKn4qHvxa7cf6l0vbtPxASEZN6mn4WlwffE5IQJoK9SmXERyU4LfvPVTgtG87JkADKzpwB395fLTj0KHHT3yXjdGGFhacUABHPbwS51rgQ8H38sPC2wXn72uSd2frB+e+d79Zlc7oWFacF3b/5279abF68+UIebSzZZ323WBbB6t9IulgxA7n7UzO4zs5eAX7r76/HU9wh7iyebH00De3NuZ9P5YP32zg87tnbmdf887y3V6d2K1EfVITh3Pw/sN7MHYiA6Hud7dprZpnjNRw1oZ1O6a/aDq//9b//zl1Z7n0oLYL/1rb+ZWe19ZWWcWxvvmr3EOn8n0+f+YPZXFY+/v64XgM/fmgTg1q17cNi4mjaKNIOa5oDc/W1CIHrczP6OUFtoMteWiaxRBh0YHffefedDtVz/XsxuunvLHRWvv3I5JOLcuzncr/Q7MO0TKW2gagAys3sJJRsAJtz9p8BPzewpM3uCEIgu5NtEkbXHjJprgCVpuktd//wvwp9/8ZVw/tmjv0EbFa99S+yMu1pttRPukpUQYhLCSeCu+LXfzP4XhPkhd3+GkMF22MzubEhrRUTWjl7qG4Bqvp+Z9ZnZ5aSgs5kNmNmpmHHcZ2bnzGwwfo2UV58xs65YFCB9bDhWnBk2sxPxs6fKPpO5ik011XpAe4DBdA8nlsG5091/D+Dufx3nggaA1yvfRkSkZU0WtMvxDKEyzCAhi7hE3F4hBqWJJPvYzCZYXHFmb4Vjp+K2DcPMb+FAvEcf8AwwkqGNy6oWgL4HjJnNjUZ3A4eT4JOISQgKPjla6wskV7AxV2/83JkMn2nZjboOfqXoFkgT6nP3Q2Z2lhCA+lLLXQYI5c66gIHyZTAxQI2T2vYmlvYZS33+cPx+DObWc07Xu3RPtTTsCUK2231AV0xEEKmk9/bbb39o2/ZtNV186eIlALZt31bTJP6li5e4du3aylsnDaUFsA2V9HjSdhMKBTzDwio0iQF3H40BCpirjED5z2XHK+1esCrLZsHFVGyRqrZt38Y3nnyypmt/9OqrAJmuf/edd1fcNpEWNkLoybyVOtYTA0xPHJLrix2KpDZnV/xzUUCJQ23jFY53kbHoaS0aWg1b8pdxOKw3fuZMhke07FCYSDNIB4M4TzNC7Omkq2PH4NMF9BBKnUEYzXo+Xjtd4fYDzJdZKz/+VoXjq6IA1HpqHg7TUJjIqvVmTB6oei+WKUYaA8pRQsBJgsrJOEfTRahyPdfTIfSO9qQ+NxLPz2XPJfM6McNtHzCcbPcQj/fF4wuy5upBAagF1Toc1uxDYev9X3cas50ZenC9UHuP77Z1vb1Ox9WVtU7awGQO96t6zxgU9pQdO5Q6V17MOZ2AsCf1mRKwoGJ2TFZYVLczDt/tLj9eDwpA0rSM2c4Ou7WpXhUJyk29f4tZLQiVJbTTgtG8NCQAxS7cQPzxQeCpVPduGDhH2AdoLzBdrXq2SNqG2zrqVpGg3P/+0cd8cl0RSCQvS1ZCqJWZfTduULfU+S5gl7s/Hye/jhF24ksbAc4TJsgUfERE2sCqAlCsgvDvgB2xSGklu4gTY9EY0JfKQX/L3c3dNyfZGdL8pl7cz9SL+4tuhuTk7a/dc6aOk+/SplY1BBerIFTdCyimCaYnzXri8Zn0delcdWlvP3xtEoDvfL230HZIPiptLyLtKVMAikNtu4Epdz9iZo/E7yerfa4ssDzB/A6oAD0xZXA8zgcda8dApNXjIs3lsaePv0AO1bBff2Hv03W+55pVcwAys8NAHyF4/AA4QlgxO0ZILKjlHl2EmkVzKX3pYbe4oOoUZemB8dwQMBR/HHX30VqeqVRdEclJLzWs3cl4v5rExK7TQH9MqU7+fj0L7EvXbEvWALn7ngqff6rSvPsSn0kSxnYQRrKOxWftTn1mUe25arL0gAbd/f74IIMwBJeuJ1SDYcpy2NMLnlKVXBeJAaemoLPg/sx2dphv+sI9X6h5sSXUvjhz6rfOrM9mbZaItIbJeg0lJkOTNSoR/vE/V/mAkGk8UaFgaKXK1zPAWJVg0ZBq2VkC0Edm9kV3/yfA40PvA2raktvMDhLKhc8kQScOvQ0D/VkandWGDRtyq1P2f468wSfXbqy4ba1g2v7Nzivvb5qbuwF4+FLoFL6aOpb4NJ77YYVzAL8tO3/p/W3csM/srFuDJZNKQ03fXP8HvQDPVfhLcyudvfFzi85Flc631dBTHQwQeiBdMBcAoKxcTqXK16nPVyq509Bq2Vmy4PYC/2hmbxCK2R0Gfgl8c7kPxhIPJ1OJB8maoHHmf5G56zK0KTNlZ9XfTfts58c3NuR2/49vbOCmfbazlmvf+/FLvPfjl3JrS5vqpf5zHY28f6sqEZO6CIVFHyQGhJSBOKfeVXZ8d4Vrl/xMXtWya+4BxXHGbjN7nPBLzxDW91TtAcVoeiJ+nxwuEQOSmZVi72gG2JEec5TmsGn2ysYNt2Z48vJ85fcbN38HwJOXFweDn93YCsBXK5wD+PubD8fPvgnA3938I657x8a6NlpqtuXm+zs3+HW+efm/LTr37Pt/tejYzzb9CQD7p35S8X4vb/42wNz9Xt78ba7bBvVws5sBtiT13MzsUFKWB5atfN1XNndEalSqYdWysyQhfNfdj7j7TyucexjA3d8sPxd/SSs/njo/wXxRPWlCt/nNjg6/mdv9O27d4Da8I7cHSFUb/Hrner9eUw90JfK8dyuKf9nPJFstUHnPH1ii8nW6YnY0lEoGa2i17CxzQLuBI2Z2b9k23YfjOTezl9z977M2Qprf7Lrb2HJgvkeTDHOmjyVui/NsW558ruK9botzP1u+HpaYzT73GtxSSZwi3bQNV9NbTyeLUCttR/1KzCBdaqvq9+LcT3L+5W/9zUy92lmA3ozJA1XvRbZq2AClOP9yEJhJJYVVrHxN6L0citcOEobtpoqqlp0lAJ03s38hJCNsAg66+z8QUqP73f1CPK8AJFUtl+ChBaitrYUWoE7mcL+q9yyvhu3u++Kf5VVkqlW+TickFFotO0sAeoTUnE8MNv8AbE71iGrKiBMRaXbK2lu9LAHIgFsAZnbnEtd0rbZB9bZp9srGDZ/Mzg0J3fh//wywZCbcjStbq54v1/nJH7FBE+SF2DR7ZeOGT2/y3o9/NHfs+vsh+aFSJtz1T++L5/5nTffv/PTfs8HXr5l3m3G3W9COt7LGZQlA+4F3zGyKEIxmzOyX8c+kEOmaq2StCfLWlf+7vcltsJbebe9nNnQ8tH3LHTVdnHX/o4tTH/PpdS2qlsbJkoY9RkjD3lSeem1mfxWv+es6t68u0hPk1SbHYfkJ8kX3XuMT5Mv9vs1udt167v5P873VpOeTPpbYEPcDuvvPvlzbvY/+Jvb5V6fW/YdqsX3LHQvuV+33zbr/0Ss//zUX3vv9ouPOrY0GHeme1H/s2NQLS/auqp2r4J5OB0W+NpS5GnaldT9rNfBIcVo14LUjgw6MjvTOs7+++z8AcC8s6l1l7XmVfge2dv8NJznKWg27l7C/zwLu/nK9GiQia49Ztp4U1H79s0d/gysAtaWaS/GY2Q+AN4FHCfXbdgHfo6y4qIiISC2y9IAed/duADP7pbvvj9+/kUvLRESkpWVNw07MmNnn3P0KFYbk1rJWn5v446sTO+/653GmXgypxtXSzrOmnPdfusg9N9at6ZpdlSbjW8UfX53YeVfp+oI08nqmnT8wdZU/nFVNNmmcLAFozMwejvXeRoG3zewycD6fpjW3SplDy+iF1WcObb55pbPz5ifA9mXv8NXPfVBj04LOTz5klo2q21WQzTevdHbe+Fcgn1fQ+ellZmusOi5SD1nSsPenvh8zs93AfZUKkMp85lBeG+FduLB05tDVjXfxhRrTzrO49F//HD75eNX3kZW7+pnN3J9T2vnbf3sErl2rQytFapOlGvY3CTvoXYjZcC8Dm8xsyN3/Ma8GNjMz4xtPfqOma7NuhPf9517DlTokIk0syxDcoVS69WnCRnJvEqqy3l/tg7FaarIJ3YOEfchn4rkeYJCwJUMfMFq++ZFIPdRzQaiIrF7mJAQzewAwdz8Sf15yr594votQxDTZY2KQEMCSbbhH3H13PFcipHiXbx8rIiItJsuW3KW4988J4l4UZnYfYVe+anYR9p9IjAF9ZtaV7DmRiCXA92Zok4iINKksAWgPYRe9YXf/y3ishzAUt6RYQy69WLUnHp8hDLnNlH+mPDCJiEjryZIF9xFhx7z0sdM1fja95fYTQLJ50qI9x4FpKmzrYGZDhM3vIMwTjdby7HbX6uue2l0rr3uS1pclC24aqJR2VXL3B2u8RxfQl8z5ZBEDjoKOiEiLyJKE0F/282ZCssC/ZLjHMAuH4yr1drpZfl5JRESaXJYhuPKKB+eBfbEW3JHlPm9mBwmp3DNm1hXngCaoMAwXkxEKUes6HBERWZ0s1bDvrPDVS0wqWOazg8DJ1PqeAVgcaGLywfFa2yQiIs0ryxDcBcIcUHrdzxTw02ofikHlRPw+OVxifvvuPbF3NEGYH9IaIBGRNpBlCK5SxlotnyuxMGhVOp9kxY2t5BkiItJ8sqwDEhERqRsFIBERKYQCkIiIFEIBSERECpElCw4zexjYDUy5+xEzeyR+P5lH40Sk+WjbC6lVlnVAhwlVrccI9dwAximrDyciIlKLLD2gQXe/H+b3AHL3j2J9N2lBP3xtEoDvfL230HZIPp7/Rfjz4FeKbYe0rywB6CMz+6K7/xOxKGncD+ijXFrWZlQCSETaTZYAtBcYN7O3gK44JDfIwuKiIrJGaC5G1roslRBKQLeZPU6o/zZD2GpbPSAREcksUxYcgLtXrf0mxfrVZ7+8cXrDdt6I8zer8dtLV4H5uaCPbx+k2y5uzLyZk9RFeLfbOPWL+tzv3anwZzIXdGXjE3R3XNL7lYapGoCqbEKXVvOGdJK/qfXbOqbWbeWOHO794bqt+HrvyOHWUoPwbj/P53K6/4frPo+vR+9XGma5HlCyCZ0BB4GzLCwYuoewMZ2sIXfd+oDvfP3Lq75PeRbc9597bdl/jUi+7rr1Pge/8od1uVd5FtyzR9/X+5WGqroOyN3Px68S0O/uR1PHzrv788S9faoxsz4zO1vh+LCZDZlZV/xzcOW/ioiINJMspXh2mNmfpg/EDel2VPuQmSUBqm+JS0YIu6t2ufvJJa4REZEWkyUJYQg4bWanCBvK7QAeYZk0bHcfgwWb0aW95e5L7hUkIiKtK0sa9kkz6yYEnC7CDqZ76pGGbWZ97j6x2vtIfakCQmtTBQQp2nJZcL1JoVEzuxO4BRwru+ZOd//9Cp/fE4foxs1sGDimQCQi0h6WDECxzM4E8/NEF1ickm3x2JaVPDwmMSTPGwFOscSckpkNEYYBAUbdfXQlzxQRkbVhyQDk7udJJSm4e3e9H25mXe4+E+9fMrOeKu0ZBRR0RERaRJbtGHorHNsU9wjKLA69nV7JZ0VEpPllScNetO9PTEAYXuGzx4HDyQ9xDZDSsEVE2sSyWXBm9hhhrqcnfp/oBnay9Pqe5PMDyTUx0eCUu4+5+4yZlczsIKGw6Q53V2VtEZE2UUsa9qPxT0t9n5ipcGyBuA5oDHi+wrkJQqKDiIi0mWUDkLvvh7CQNPleRERktbIsRE0CUS9hP6BSskZIREQkq5oDkJltImStbSLUbusxs8vAI6tYiCoiIm0qSxbcKHDY3e9390fdfSchM+5EPk0TEZFWliUA9ZXvhhoXhy65eFRERGQpWQLQR2Z2T/pAnA9adTFSERFpP1m2Y9gLvG1mx4jrdgjbMTySQ7tERKTF1dwDiruiJgVKjVA4tEeZcCIishJZhuAA+gk9nw/d/Siwq1KNOBERkeVkKUb6A+AQoefzRDw8ToUacSIiIsvJMgf0uLvfD2Bxf213/8jMuvJomIiItLasWXBfjN87zG1apyw4ERHJLGsW3LiZvQV0mdlhYBBQBWsREcksSy24EtBtZo8TFp/OALvinkBVmVkfcNTd+8uO9xCC2ARhy4bRZIdUERFpbVUDkJndWeHwqfJrqtWCi/sBTVN536ARd98drysRNrfbt1yjRUSk+S3XA7pAmO+x1DFPfW/x5y1L3SDuB0TMW5j/YOj9pK8rmdleFIBERNpC1QDk7t05PruPMIy3gJn1xOE+ERFpYVnWAT1cvujUzB4zs4dX+OxKwW0a6Frh/UREpIlkScMeAcp7JqcJ8za5M7MhMxuPX0ONeKaIiOQnSxr2lvJkg1UuRK3U2+mmwrBcfNYoYU8iERFpAVl6QONm9qfpA6vcjmGCCsNwmv8REWkPWXpA+wlB6ByhBtwOQnHS/qqfWkLMepv7OWbFHV/JvUREpPmsZCHqIGFbhrHyHVIrieuA+uL3w8CpJDUb2GNmB4kLUd1dKdgiIm0iSw8IAHc/mfH6MWAMeL7CuVLq+Fj5eRERaV01ByAzm2bhItREyd0frF+TRESkHWTpAZXP9WwGhoD/W7/miIhIu8gyB3S+7NB5YL+ZvQEcqWurRESk5WUZgqtUmLQnfomIiGSSZQjuAosLk04By2bCiYiIlMsyBJdnYVIREWkzWSohiIiI1M1yG9LVVOna3d+sT3NERKRdLDcEt2jxaIoTion2AB31apCIiLSH5Tak21XpuJltAp4hrANSEoKIiGSWaQ7IzO40s5cIWylsAvrdfW8uLRMRkZZWUwAys3vN7Dhhrx4Hut39LyssThUREalJ1QBkZr2x0sFZ4BywOQaele4BJCIiAiyfhDBBCD6HCMNuj6T38Em4++v1b5qIiLSy5QLQacKQW7V5HgdWHIDiHkHnCJvR7QWms275ICIizWe5LLjdDWrHCDAMHHb30QY9U0RECpR5Q7ocvOXui8f1RESkpa2ZUjxm1ld0G0REpHHWQgDqMbMBoGRmw0sFIjMbMrPx+DXU4DaKiEidFT4E5+5z5X7MbAQ4BeyocN0ooPkhEZEWUXgPyMy6ku/dvYQ2uBMRaQuFBqA49Ha6yDaIiEgxiu4BjQOHkx/MbBDQGiARkTZQ6ByQu8+YWcnMDhLqzO1w9z1FtklERBpjLSQhTBBK/oiISBspeghORETalAKQiIgUQgFIREQKoQAkIiKFUAASEZFCKACJiEghFIBERKQQCkAiIlIIBSARESmEApCIiBRCAUhERAqhACQiIoVQABIRkUIUXg3bzHqAQUJF7D5g1N1nCm2UiIjkrvAABIy4+24AMysBw8C+YpskIiJ5K3pL7p70z+5eAvYW1BwREWmgontAfYSdUBcws54YjOpiyu7i+8+9Vq/bLbp3Nx+0xXPX5u+6lWeP/ian527Vcwt+rrQ2c/fiHm42BOxOb8NtZueAPXGn1PJrh+KPo+4+mme78rz/WnpuO/2uem7rP1eaS9Nkwbn7qLvvil95/w97aPlLWua57fS76rmt/1xpIkUHoGmgq+xYNxWG5UREpLUUHYAmCAFngXrO/4iIyNpUaAAqDzQxK+54Qc1JK2rsuojnttPvque2/nOliRSahACLF6K6+/OFNkhERBqi8AAkIiLtqeg5IBERaVMKQClmNmBmbmYDDXjWQTM7l/q6XPbzqQa24XL8Gs77mannnk39zo16rt5vY57Z8HcrzUlDcClmdoKQFj6TXhzboGc70F++ADfH550ipLs/5e4zZtYFDLj7yZyfm/w33ufupTgHeMjdc6//p/eb7/st8t1Kc1IASolVGHYD59zdGvzshv0FFatK7HP3/ryfVfbcQWDY3Xc08rmp5+v95vfMQt+tNCcNwUXx/0BjMTW8FH9uVYeAwwU89xlCtfOG0/vNXWHvVpqXAtC8fcCJ+P0I8ESBbclbDyHtvdH6gPECngt6v3kr8t1Kk1IAAuL4+C53H4uHTgKD8bjU10yjH6j32zAzRTdAmosCULAX6IrZO2eZ/5dyq+5NVCL8i7XRZgj/OgfAzEZippSX7w1VZ3q/+ZuhmHcrTUwBKNhDmCCe+yKMozc0U6qBRghj9o02SvjvCoC774uT1jM5P1fvN39FvVtpYm0fgFLDM+Vj5ieBgVYcpknKHZnZifTvZ2YHc37uIaAn/uu4a7nr60HvtzHvt4h3K82v7QMQYd+SRQVQY7bUBC26r0nsBZSAZNHg5QY9N/lX8dxiRcJ//+mcHqn326D3W8C7lSandUAiIlII9YBERKQQCkAiIlIIBSARESmEApCIiBRCAUhERAqhACQiIoVQABIRkUIoAImISCH+P5rpYICCU+0XAAAAAElFTkSuQmCC\n",
      "text/plain": [
       "<Figure size 432x216 with 1 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "nucneat = {i:[] for i in letters}\n",
    "families = [i for i in families2 if i in seq]\n",
    "#Calculate mean nucleotide usage for NEAT1 orthologs\n",
    "for gene in families:\n",
    "    allen = len(seq[gene])\n",
    "    for letter in letters:\n",
    "        nucneat[letter].append(seq[gene].count(letter)/allen*100)\n",
    "nval = [np.mean(nucneat[letter]) for letter in letters]\n",
    "nerr = [np.std(nucneat[letter]) for letter in letters]\n",
    "\n",
    "#Calculate mean nucleotide usage for MALAT1 orthologs\n",
    "nucmal = {i:[] for i in letters}\n",
    "families = [i for i in families2 if i in seq2]\n",
    "for gene in families:\n",
    "    allen = len(seq2[gene])\n",
    "    for letter in letters:\n",
    "        nucmal[letter].append(seq2[gene].count(letter)/allen*100)\n",
    "mval = [np.mean(nucmal[letter]) for letter in letters]\n",
    "merr = [np.std(nucmal[letter]) for letter in letters]\n",
    "\n",
    "#Calculate mean nucleotide usage for CDSs of protein-coding genes\n",
    "filelabel = 'CDS'\n",
    "a = []\n",
    "err = []\n",
    "for letter in letters:\n",
    "    inter = [float(np.mean(val[filelabel][gene][letter])) for gene in val[filelabel]]\n",
    "    a.append(np.nanmean(inter))\n",
    "    err.append(np.nanstd(inter))\n",
    "    \n",
    "#plot first four columns\n",
    "pos = [i for i in range(len(a))]\n",
    "\n",
    "plt.figure(figsize=(6,3))\n",
    "plt.bar(pos,a,yerr=err, color='#e7e9e8',edgecolor='#6c6c6c',\n",
    "        error_kw={'linewidth':0.5,'linewidth':1.5,'ecolor':'#6c6c6c'},linewidth=1.5,label='CDS')\n",
    "\n",
    "plt.bar(pos,nval, color='None', fill=False, edgecolor='#ce591e',linewidth=1.5)\n",
    "plt.errorbar([i - 1/4 for i in pos], nval, yerr=nerr, linestyle='none', color='#ce591e')\n",
    "\n",
    "plt.bar(pos,mval, color='None', fill=False, edgecolor='#516fa9',linewidth=1.5)\n",
    "plt.errorbar([i + 1/4 for i in pos], mval, yerr=merr, linestyle='none', color='#516fa9')\n",
    "\n",
    "#Calculate mean nucleotide usage for UTRs of protein-coding genes\n",
    "filelabel = 'UTR'\n",
    "a = []\n",
    "err = []\n",
    "for letter in letters:\n",
    "    inter = [float(np.mean(val[filelabel][gene][letter])) for gene in val[filelabel]]\n",
    "    a.append(np.nanmean(inter))\n",
    "    err.append(np.nanstd(inter))\n",
    "pos = range(4,8)\n",
    "\n",
    "#Finish the plot\n",
    "plt.bar(pos,a,yerr=err, color='#eed1a8',edgecolor='#6c6c6c',\n",
    "        error_kw={'linewidth':0.5,'linewidth':1.5,'ecolor':'#6c6c6c'},linewidth=1.5,label='UTR')\n",
    "\n",
    "plt.bar(pos,nval, color='None', fill=False, edgecolor='#ce591e',linewidth=1.5,label=r\"$\\mathit{NEAT1}$\")\n",
    "plt.errorbar([i - 1/4 for i in pos], nval, yerr=nerr, linestyle='none', color='#ce591e')\n",
    "\n",
    "plt.bar(pos,mval, color='None', fill=False, edgecolor='#516fa9',linewidth=1.5, label=r\"$\\mathit{MALAT1}$\")\n",
    "plt.errorbar([i + 1/4 for i in pos], mval, yerr=merr, linestyle='none', color='#516fa9')\n",
    "\n",
    "\n",
    "plt.ylabel('Nucleotide usage, \\%',fontsize=14)\n",
    "plt.box(False)\n",
    "\n",
    "plt.legend(bbox_to_anchor=(1.1, 1),fontsize=10,frameon=False)\n",
    "plt.xticks(range(8),letters+letters,fontsize=14)\n",
    "plt.yticks(fontsize=14)\n",
    "plt.tight_layout()\n",
    "\n",
    "plt.savefig('./FiguresPublication/NuclUsageOrthologNEAT1MALAT1.pdf',dpi=300)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bf9cc4bb",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
