{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "a525d86c",
   "metadata": {},
   "outputs": [],
   "source": [
    "import re"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "829f722f",
   "metadata": {},
   "outputs": [],
   "source": [
    "#To build a reverse complimentary strand\n",
    "def rc(seq):\n",
    "    complement = {'A': 'T', 'C': 'G', 'G': 'C', 'T': 'A','N':'N'}\n",
    "    return(\"\".join(complement.get(base, base) for base in reversed(seq)))\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "31af9b58",
   "metadata": {},
   "source": [
    "### Write into FASTA file sequences of identified orthologs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "6f779122",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Collect contigs names and coordinates, add an offset - extra sequence to both borders\n",
    "offset = 200\n",
    "d = {}\n",
    "with open('./Key_Files/ManualCuration') as file1:\n",
    "    for line in file1:\n",
    "        if not line.startswith('#'):\n",
    "            dat = line.strip().split('\\t')\n",
    "            d[dat[0]] = {}\n",
    "            if dat[0] in ['GCA_003118495.1_macFus_1.0','GCA_002335545.1_Aspe_assembly01']:\n",
    "                start = int(dat[2])\n",
    "                stop = int(dat[3])\n",
    "                d[dat[0]]['neat'] = [dat[1],start + offset,stop,'minus']\n",
    "\n",
    "            else:          \n",
    "                if dat[4] != 'Nan':\n",
    "                    start = int(dat[2])\n",
    "                    stop = int(dat[3])\n",
    "                    if start > stop:\n",
    "                        d[dat[0]]['neat'] = [dat[1],start + offset,stop - offset,'minus']\n",
    "                    else:\n",
    "                        d[dat[0]]['neat'] = [dat[1],start - offset,stop + offset,'plus']\n",
    "            if dat[8] != 'Nan':\n",
    "                start = int(dat[6])\n",
    "                stop = int(dat[7])\n",
    "                if start > stop:\n",
    "                    d[dat[0]]['malat'] = [dat[5],start + offset,stop - offset,'minus']\n",
    "                else:\n",
    "                    d[dat[0]]['malat'] = [dat[5],start - offset,stop + offset,'plus']        \n",
    "                    \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "8557ea5a",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "dict_keys(['Mouse_NEAT1', 'Marsupials_NEAT1', 'Human_NEAT1', 'Mouse_MALAT1', 'Human_MALAT1'])\n"
     ]
    }
   ],
   "source": [
    "#Gather reference sequences\n",
    "ref = {}\n",
    "ref_names = ['NEAT1_Mouse','NEAT1_Marsupials','NEAT1_Human','MALAT1_Mouse','MALAT1_Human']\n",
    "for ref_name in ref_names:\n",
    "    seq_file = ref_name + '.fasta'\n",
    "    with open(seq_file) as file1:\n",
    "        fl = 0\n",
    "        for line in file1:\n",
    "            if line.startswith('>') and fl == 0:\n",
    "                l = []\n",
    "                name = line.rstrip().lstrip('>')\n",
    "                fl = 1\n",
    "            elif line.startswith('>') and fl == 1:\n",
    "                ref[name] = ''.join(l)\n",
    "                l = []\n",
    "                name = line.rstrip().lstrip('>')\n",
    "            elif not line.startswith('>'):\n",
    "                l.append(line.strip())\n",
    "        ref[name] = ''.join(l)\n",
    "print(ref.keys())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "f5c06795",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Create a single smaller file with contigs on which orthologs were identified\n",
    "\n",
    "out = open('NeededContigsAllGenomes.fasta','w')\n",
    "for item in d:\n",
    "    seq_file = './downloads/' + item + '_genomic.fna'\n",
    "    acc = []\n",
    "    if 'neat' in d[item]:\n",
    "        acc.append(d[item]['neat'][0])\n",
    "    else:\n",
    "        acc.append('Nan')\n",
    "    if 'malat' in d[item]:\n",
    "        acc.append(d[item]['malat'][0])\n",
    "    else:\n",
    "        acc.append('Nan')\n",
    "    with open(seq_file) as file1:\n",
    "        fl = 0\n",
    "        for line in file1:\n",
    "            if line.startswith('>') and fl == 0:\n",
    "                l = []\n",
    "                name = line.lstrip('>').split(' ')[0]\n",
    "                fl = 1\n",
    "            elif line.startswith('>') and fl == 1:\n",
    "                if name in acc:\n",
    "                    out.write('>' + item + '|' + name + '\\n' + ''.join(l).upper() + '\\n')\n",
    "                    \n",
    "                l = []\n",
    "                name = line.lstrip('>').split(' ')[0]\n",
    "            elif not line.startswith('>'):\n",
    "                l.append(line.strip())\n",
    "        if name in acc:\n",
    "            out.write('>' + item + '|' + name + '\\n' + ''.join(l).upper() + '\\n')\n",
    "out.close()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "861fd178",
   "metadata": {},
   "outputs": [],
   "source": [
    "out_neat = open('neat1_seqs_unrefined_offset_200.fasta','w')\n",
    "out_malat = open('malat1_seqs_unrefined_offset_200.fasta','w')\n",
    "\n",
    "out_malat.write('>Mouse_MALAT1\\n' + ref['Mouse_MALAT1'] + '\\n')\n",
    "out_malat.write('>Human_MALAT1\\n' + ref['Human_MALAT1'] + '\\n')\n",
    "\n",
    "for item in ['Mouse_NEAT1', 'Marsupials_NEAT1', 'Human_NEAT1']:\n",
    "    out_neat.write('>' + item + '\\n' + ref[item] + '\\n')\n",
    "\n",
    "# function to properly cut the gene depending on the strand\n",
    "def processing(lib, accession, seq):\n",
    "    global d\n",
    "    # Cut out NEAT1 gene\n",
    "    if 'neat' in d[lib] and accession == d[lib]['neat'][0]:\n",
    "        if d[lib]['neat'][3] == 'plus':\n",
    "            if (d[lib]['neat'][2] + 1) <= len(seq):\n",
    "                cutout = seq[d[lib]['neat'][1]:d[lib]['neat'][2] + 1]\n",
    "            else:\n",
    "                cutout = seq[d[lib]['neat'][1]:]\n",
    "            cutoutN = re.sub('N+','NN',cutout)\n",
    "            out_neat.write('>' + lib + '_NEAT1\\n' + cutoutN + '\\n')\n",
    "        elif d[lib]['neat'][3] == 'minus':\n",
    "            if (d[lib]['neat'][1] + 1) <= len(seq):\n",
    "                cutout = seq[d[lib]['neat'][2]:d[lib]['neat'][1] + 1]\n",
    "            else:\n",
    "                cutout = seq[d[lib]['neat'][2]:]\n",
    "            cutoutN = re.sub('N+','NN',cutout)\n",
    "            out_neat.write('>' + lib + '_NEAT1\\n' + rc(cutoutN) + '\\n')\n",
    "        else:\n",
    "            print('NEAT1',lib)\n",
    "    # Cut out MALAT1 gene\n",
    "    if 'malat' in d[lib] and accession == d[lib]['malat'][0]:\n",
    "        if d[lib]['malat'][3] == 'plus':\n",
    "            if (d[lib]['malat'][2] + 1) < len(seq):\n",
    "                cutout = seq[d[lib]['malat'][1]:d[lib]['malat'][2] + 1]\n",
    "            else:\n",
    "                cutout = seq[d[lib]['malat'][1]:]\n",
    "            cutoutN = re.sub('N+','NN',cutout)\n",
    "            out_malat.write('>' + lib + '_MALAT1\\n' + cutoutN + '\\n')\n",
    "        elif d[lib]['malat'][3] == 'minus':\n",
    "            if (d[lib]['malat'][1] + 1) < len(seq):\n",
    "                cutout = seq[d[lib]['malat'][2]:d[lib]['malat'][1] + 1]\n",
    "            else:\n",
    "                cutout = seq[d[lib]['malat'][2]:]\n",
    "            cutoutN = re.sub('N+','NN',cutout)\n",
    "            out_malat.write('>' + lib + '_MALAT1\\n' + rc(cutoutN) + '\\n')\n",
    "        else:\n",
    "            print('MALAT1',lib)\n",
    " \n",
    "        \n",
    "#Parsing the FASTA file with contigs and cut the orthologs\n",
    "\n",
    "with open('NeededContigsAllGenomes.fasta') as file1:\n",
    "    fl = 0\n",
    "    for line in file1:\n",
    "        if line.startswith('>') and fl == 0:\n",
    "            l = []\n",
    "            name = line.lstrip('>').split('|')[0]\n",
    "            accession = line.strip().split('|')[1]\n",
    "            fl = 1\n",
    "        elif line.startswith('>') and fl == 1:\n",
    "            processing(name, accession, ''.join(l).upper())\n",
    "            l = []\n",
    "            name = line.lstrip('>').split('|')[0]\n",
    "            accession = line.strip().split('|')[1]\n",
    "        elif not line.startswith('>'):\n",
    "            l.append(line.strip())\n",
    "    processing(name, accession, ''.join(l).upper())\n",
    "out_neat.close()\n",
    "out_malat.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "98ebb7ca",
   "metadata": {},
   "source": [
    "### Multiple Sequence Alignment and preparation for it\n",
    "\n",
    "To identify more accurately gene coordinates and structural features we performed MSA of NEAT1 orthologs in groups, which would be big enough and close phylogenetically as we expect high divergence of the primary sequence"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "id": "deb22b6f",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Defining the groups\n",
    "groups = {'MarsMono':['MONOTREMATA','DIDELPHIMORPHIA','MICROBIOTHERIA','DIPROTODONTIA','DASYUROMORPHIA'],\n",
    "'EuPerisPholi': ['EULIPOTYPHLA','PERISSODACTYLA','PHOLIDOTA'],          \n",
    "'AfroSirTub':['MACROSCELIDEA','PILOSA','PROBOSCIDEA','AFROSORICIDA','CINGULATA','SIRENIA','TUBULIDENTATA','HYRACOIDEA'],\n",
    "'RodLagoScan':['LAGOMORPHA','RODENTIA','SCANDENTIA'],\n",
    "'PrimDerm':['PRIMATES','DERMOPTERA'],\n",
    "'Art':['ARTIODACTYLA'],'Chirop':['CHIROPTERA'],'Carn':['CARNIVORA']}\n",
    "groups_r = {}\n",
    "for group in groups:\n",
    "    for item in groups[group]:\n",
    "        groups_r[item] = group\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "8d13f802",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Retrieve taxonomy class information for the genomes\n",
    "tax = {}\n",
    "with open('./Key_Files/Metadata') as file1:\n",
    "    for line in file1:\n",
    "        if not line.startswith('#'):\n",
    "            dat = line.strip().split('\\t')\n",
    "            tax[dat[0]] = dat[7].upper()\n",
    "       "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "id": "3b48a7cc",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Mouse\n",
      "Marsupials\n",
      "Human\n",
      "Mouse\n",
      "Human\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "[None, None, None, None, None, None, None, None]"
      ]
     },
     "execution_count": 52,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#Write sequences in separate FASTA files together with the most relevant known ortholog\n",
    "order = [group for group in groups]\n",
    "handler = [open('./MSA/RAW/NEAT1_' + group + '.fasta','w') for group in order]\n",
    "\n",
    "for i in range(len(order)):\n",
    "    if order[i] == 'MarsMono':\n",
    "        handler[i].write('>Marsupials_NEAT1\\n' + ref['Marsupials_NEAT1'] + '\\n')\n",
    "    elif order[i] in ['PrimDerm','Art','Carn','EuPerisPholi']:\n",
    "        handler[i].write('>Human_NEAT1\\n' + ref['Human_NEAT1'] + '\\n')\n",
    "    elif order[i] == 'RodLagoScan':\n",
    "        handler[i].write('>Mouse_NEAT1\\n' + ref['Mouse_NEAT1'] + '\\n')\n",
    "    else:\n",
    "        for item in ['Mouse_NEAT1', 'Human_NEAT1']:\n",
    "            handler[i].write('>' + item + '\\n' + ref[item] + '\\n')\n",
    "\n",
    "\n",
    "with open('neat1_seqs_unrefined_offset_200.fasta') as file1:\n",
    "    fl = 0\n",
    "    for line in file1:\n",
    "        if line.startswith('>') and fl == 0:\n",
    "            l = []\n",
    "            name = '_'.join(line.lstrip('>').split('_')[:-1])\n",
    "\n",
    "            fl = 1\n",
    "        elif line.startswith('>') and fl == 1:\n",
    "            if name in tax:\n",
    "                idx = order.index(groups_r[tax[name]])\n",
    "                handler[idx].write('>' + name + '\\n' + ''.join(l) + '\\n')\n",
    "            else:\n",
    "                print(name)\n",
    "            l = []\n",
    "            name = '_'.join(line.lstrip('>').split('_')[:-1])\n",
    "        elif not line.startswith('>'):\n",
    "            l.append(line.strip())\n",
    "    if name in tax:\n",
    "        idx = order.index(groups_r[tax[name]])\n",
    "        handler[idx].write('>' + name + '\\n' + ''.join(l) + '\\n')\n",
    "\n",
    "[i.close() for i in handler]\n",
    "\n",
    "handler = [open('./MSA/RAW/MALAT1_' + group + '.fasta','w') for group in order]\n",
    "for item in handler:\n",
    "    item.write('>Mouse_MALAT1\\n' + ref['Mouse_MALAT1'] + '\\n')\n",
    "    item.write('>Human_MALAT1\\n' + ref['Human_MALAT1'] + '\\n')\n",
    "with open('malat1_seqs_unrefined_offset_200.fasta') as file1:\n",
    "    fl = 0\n",
    "    for line in file1:\n",
    "        if line.startswith('>') and fl == 0:\n",
    "            l = []\n",
    "            name = '_'.join(line.lstrip('>').split('_')[:-1])\n",
    "\n",
    "            fl = 1\n",
    "        elif line.startswith('>') and fl == 1:\n",
    "            if name in tax:\n",
    "                idx = order.index(groups_r[tax[name]])\n",
    "                handler[idx].write('>' + name + '\\n' + ''.join(l) + '\\n')\n",
    "            else:\n",
    "                print(name)\n",
    "            l = []\n",
    "            name = '_'.join(line.lstrip('>').split('_')[:-1])\n",
    "        elif not line.startswith('>'):\n",
    "            l.append(line.strip())\n",
    "    if name in tax:\n",
    "        idx = order.index(groups_r[tax[name]])\n",
    "        handler[idx].write('>' + name + '\\n' + ''.join(l) + '\\n')\n",
    "\n",
    "[i.close() for i in handler]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "id": "fe20bc2b",
   "metadata": {},
   "outputs": [],
   "source": [
    "#In command line MSA was performed\n",
    "\n",
    "#for i in `ls *.fasta`; do mafft \"$i\" > ./MSA/msa_\"$i\"; done"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9b32db8a",
   "metadata": {},
   "source": [
    "MSAs (made by MAFFT) of approximately cut sequences were visually explored in AlignmentViewer. 'TATA'-boxex together with 'triple helix' patterns were identified. Coordinates of these features were manually checked and put in the dictionary below. Precise gene sequences and structural elements were retrieved. "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4f632c78",
   "metadata": {},
   "source": [
    "### Retrieval of final sequences of orthologs and their gene features"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "69febcb7",
   "metadata": {},
   "outputs": [],
   "source": [
    "coor = {'NEAT1':{'MarsMono':{'gene':[486,63741], 'trna':[63741,63805],'TATA':[438,486],'promoter':[240,443],'triplex':[63643,63741]},\n",
    "'EuPerisPholi': {'gene':[346,40815], 'trna':[40815,40880],'TATA':[303,346],'promoter':[210,308],'triplex':[40727,40815]},          \n",
    "'AfroSirTub':{'gene':[327,42714], 'trna':[42714,42780],'TATA':[289,327],'promoter':[80,294],'triplex':[42623,42714]},\n",
    "'RodLagoScan':{'gene':[490,120889], 'trna':[120889,120965],'TATA':[448,490],'promoter':[280,453],'triplex':[120791,120889]},\n",
    "'PrimDerm':{'gene':[349,68528], 'trna':[68528,68590],'TATA':[313,349],'promoter':[150,318],'triplex':[68439,68528]},\n",
    "'Art':{'gene':[290,67634], 'trna':[67634,67700],'TATA':[255,290],'promoter':[40,259],'triplex':[67544,67634]},\n",
    "'Chirop':{'gene':[343,40898], 'trna':[40898,40960],'TATA':[301,343],'promoter':[140,306],'triplex':[40798,40898]},\n",
    "'Carn':{'gene':[301,73252], 'trna':[73253,73338],'TATA':[265,301],'promoter':[40,270],'triplex':[73064,73253]}},\n",
    "        \n",
    "       'MALAT1':{'MarsMono':{'gene':[2748,14813], 'trna':[14813,14880],'TATA':[2714,2748],'promoter':[2600,2719],'triplex':[14726,14813]},\n",
    "'EuPerisPholi': {'gene':[2175,13550], 'trna':[13550,13610],'TATA':[2078,2175],'promoter':[1700,2083],'triplex':[13461,13550]},          \n",
    "'AfroSirTub':{'gene':[1950,10572], 'trna':[10572,10640],'TATA':[1916,1950],'promoter':[1710,1921],'triplex':[10482,10572]},\n",
    "'RodLagoScan':{'gene':[2793,15642], 'trna':[15642,15707],'TATA':[2723,2793],'promoter':[2590,2728],'triplex':[15548,15642]},\n",
    "'PrimDerm':{'gene':[1979,13170], 'trna':[13170,13240],'TATA':[1945,1979],'promoter':[1730,1950],'triplex':[13073,13170]},\n",
    "'Art':{'gene':[1332,21534], 'trna':[21534,21600],'TATA':[1298,1332],'promoter':[1150,1303],'triplex':[20318,21534]},\n",
    "'Chirop':{'gene':[1327,11403], 'trna':[11403,11470],'TATA':[1293,1327],'promoter':[1100,1298],'triplex':[11310,11403]},\n",
    "'Carn':{'gene':[1695,10552], 'trna':[10552,10615],'TATA':[1657,1695],'promoter':[1540,1662],'triplex':[10460,10552]}}}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "4ad27290",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Write in FASTA files final versions of orthologs and their gene features\n",
    "for gene in coor:\n",
    "    for group in coor[gene]:\n",
    "        msa = {}\n",
    "        with open('./MSA/RAW/MSA/msa_' + gene + '_' + group + '.fasta') as file1:\n",
    "            fl = 0\n",
    "            for line in file1:\n",
    "                if line.startswith('>') and fl == 0:\n",
    "                    l = []\n",
    "                    name = line.rstrip().lstrip('>')\n",
    "                    fl = 1\n",
    "                elif line.startswith('>') and fl == 1:\n",
    "                    msa[name] = ''.join(l)\n",
    "                    l = []\n",
    "                    name = line.rstrip().lstrip('>')\n",
    "                elif not line.startswith('>'):\n",
    "                    l.append(line.strip())\n",
    "            msa[name] = ''.join(l)\n",
    "        for feature in coor[gene][group]:\n",
    "            out = open('./FinalSeqs/' + gene + '_' + feature + '_' + group + '.fasta','w')\n",
    "            start = coor[gene][group][feature][0]\n",
    "            stop = coor[gene][group][feature][1]\n",
    "            for item in msa:\n",
    "                seq = ''.join(list(msa[item])[start:stop + 1]).replace('-','').upper()\n",
    "                if seq and len(seq) > 20:\n",
    "                    out.write('>' + item + '\\n' + seq + '\\n')\n",
    "            \n",
    "            out.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "6d00a84a",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Combine groups into a single file for each of features\n",
    "features = ['gene','trna','TATA','promoter','triplex']\n",
    "for gene in coor:\n",
    "    for feature in features:\n",
    "        out = open('./FinalSeqs/' + gene + '_' + feature + '_all.fasta','w')\n",
    "        msa = {}\n",
    "        for group in coor[gene]:\n",
    "            with open('./FinalSeqs/' + gene + '_' + feature + '_' + group + '.fasta') as file1:\n",
    "                fl = 0\n",
    "                for line in file1:\n",
    "                    if line.startswith('>'):\n",
    "                        name = line\n",
    "                    else:\n",
    "                        msa[name] = line\n",
    "        for item in msa:\n",
    "            out.write(item + msa[item])\n",
    "        out.close()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
