{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "4e642b40",
   "metadata": {},
   "outputs": [],
   "source": [
    "import csv"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3841c3c4",
   "metadata": {},
   "source": [
    "### Selection of genomic assemblies and their downloading"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9e57ed58",
   "metadata": {},
   "source": [
    "Parsing the list of genomes, obtained from NCBI FTP on June 2023\n",
    "\n",
    "https://ftp.ncbi.nlm.nih.gov/genomes/ASSEMBLY_REPORTS/assembly_summary_genbank.txt\n",
    "\n",
    "See ftp://ftp.ncbi.nlm.nih.gov/genomes/README_assembly_summary.txt for a description of the columns in this file.\n",
    "\n",
    "#assembly_accession\tbioproject\tbiosample\twgs_master\trefseq_category\ttaxid\tspecies_taxid\torganism_name\tinfraspecific_name\tisolate\tversion_status\tassembly_level\trelease_type\tgenome_rep\tseq_rel_date\tasm_name\tasm_submitter\tgbrs_paired_asm\tpaired_asm_comp\tftp_path\texcluded_from_refseq\trelation_to_type_material\tasm_not_live_date\tassembly_type\tgroup\tgenome_size\tgenome_size_ungapped\tgc_percent\treplicon_count\tscaffold_count\tcontig_count\tannotation_provider\tannotation_name\tannotation_date\ttotal_gene_count\tprotein_coding_gene_count\tnon_coding_gene_count\tpubmed_id\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "31c7bd54",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2066 622 577\n"
     ]
    }
   ],
   "source": [
    "ftp = {}\n",
    "latin = {}\n",
    "names = []\n",
    "rep = {}\n",
    "with open('assembly_summary_genbank.txt') as file1:\n",
    "    for line in file1:\n",
    "        if not line.startswith('#') and 'ExonCaptKit' not in line:\n",
    "            dat = line.strip().split('\\t')\n",
    "            if dat[20] == 'na' and dat[13] == 'Full':\n",
    "                if dat[4] == 'representative genome':\n",
    "                    rep[dat[7]] = dat[19]\n",
    "                if dat[7] not in ftp:\n",
    "                    ftp[dat[7]] = []\n",
    "                ftp[dat[7]].append(dat[19])\n",
    "                latin[dat[19].split('/')[-1]] = [dat[7],dat[6]]\n",
    "                names.append(dat[7])\n",
    "print(len(names),len(set(names)),len(rep))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "id": "6dc9ce25",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Write down the path to the assembly, 'ftp_path' column\n",
    "out = open('FTP_for_genomes','w')\n",
    "\n",
    "for item in names:\n",
    "    if item in rep and item != 'Canis lupus familiaris':\n",
    "        out.write(rep[item] + '/' + rep[item].split('/')[-1] + '_genomic.fna.gz\\n')\n",
    "    else:\n",
    "        for indi in ftp[item]:\n",
    "            out.write(indi + '/' + indi.split('/')[-1] + '_genomic.fna.gz\\n')\n",
    "\n",
    "out.close()\n",
    "\n",
    "# Run wget to download assemblies from the file\n",
    "# for item in `cat FTP_for_genomes`; do wget \"$item\"; done"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "605375ed",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Make the file (Namesystem) connecting genome IDs with species names and their taxid\n",
    "# GCF_000151735.1_Cavpor3.0\tCavia porcellus\t10141\n",
    "\n",
    "genomes = []\n",
    "with open('FTP_for_genomes') as file1:\n",
    "    for line in file1:\n",
    "        genomes.append(line.split('/')[-2])\n",
    "        \n",
    "out = open('Namesystem','w')\n",
    "for item in genomes:\n",
    "    out.write(item + '\\t' + '\\t'.join(latin[item]) + '\\n')\n",
    "out.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "50e7bab0",
   "metadata": {},
   "source": [
    "### Retrieving approximate genomic coordinates for NEAT1 and MALAT1 orthologs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "2f2398c4",
   "metadata": {},
   "outputs": [],
   "source": [
    "#### Set of commands which were run in command line outside of python\n",
    "\n",
    "# cut -f 1 Namesystem > liblist\n",
    "\n",
    "# for i in `cat liblist`; do makeblastdb -in ./downloads/\"$i\"_genomic.fna -dbtype 'nucl'; \n",
    "# blastn -query known_NEAT1_MALAT1_and_features.fasta -db ./downloads/\"$i\"_genomic.fna -task blastn\n",
    "# -out ./BLAST_features/\"$i\".txt -max_target_seqs 3  \n",
    "# -outfmt \"6 qseqid sseqid length pident qstart qend sstart send evalue bitscore\" -max_hsps 3; done"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "b1d32fdc",
   "metadata": {},
   "outputs": [],
   "source": [
    "with open('liblist') as file1:\n",
    "    libs = file1.read().split('\\n')[:-1]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "ac096add",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "GCA_000002285.4_Dog10K_Boxer_Tasha No MALAT start\n",
      "GCA_000152225.2_Pcap_2.0 No MALAT start\n",
      "GCA_000164785.2_C_hoffmanni-2.0.1 No NEAT start\n",
      "GCA_000164785.2_C_hoffmanni-2.0.1 No MALAT start\n",
      "GCA_000164785.2_C_hoffmanni-2.0.1 No NEAT stop\n",
      "GCA_000181375.1_ASM18137v1 No NEAT start\n",
      "GCA_000181375.1_ASM18137v1 No MALAT start\n",
      "GCA_000181375.1_ASM18137v1 No NEAT stop\n",
      "GCA_000181375.1_ASM18137v1 No MALAT stop\n",
      "GCA_000181415.1_ASM18141v1 No NEAT start\n",
      "GCA_000181415.1_ASM18141v1 No MALAT start\n",
      "GCA_000181415.1_ASM18141v1 No NEAT stop\n",
      "GCA_000181415.1_ASM18141v1 No MALAT stop\n",
      "GCA_000331495.1_Beagle No NEAT start\n",
      "GCA_000331495.1_Beagle No MALAT start\n",
      "GCA_000331495.1_Beagle No NEAT stop\n",
      "GCA_000331495.1_Beagle No MALAT stop\n",
      "GCA_000465285.1_ASM46528v1 No NEAT start\n",
      "GCA_000465285.1_ASM46528v1 No NEAT stop\n",
      "GCA_000688575.1_CavAp1.0 No NEAT start\n",
      "GCA_000688575.1_CavAp1.0 No MALAT start\n",
      "GCA_000688575.1_CavAp1.0 No MALAT stop\n",
      "GCA_000754665.1_Bison_UMD1.0 No NEAT start\n",
      "GCA_000754665.1_Bison_UMD1.0 No NEAT stop\n",
      "GCA_000978805.1_ASM97880v1 No NEAT start\n",
      "GCA_000978805.1_ASM97880v1 No NEAT stop\n",
      "GCA_001262655.1_Emacaco_refEf_BWA_oneround No NEAT start\n",
      "GCA_001262655.1_Emacaco_refEf_BWA_oneround No NEAT stop\n",
      "GCA_001262665.1_Eflavifronsk33QCA No NEAT start\n",
      "GCA_001262665.1_Eflavifronsk33QCA No NEAT stop\n",
      "GCA_001458135.2_marMar No NEAT start\n",
      "GCA_001675575.1_ASM167557v1 No MALAT stop\n",
      "GCA_001685095.1_ETalpinus_0.1 No NEAT start\n",
      "GCA_001685095.1_ETalpinus_0.1 No NEAT stop\n",
      "GCA_002197005.1_CerEla1.0 No NEAT start\n",
      "GCA_002335545.1_Aspe_assembly01 No NEAT start\n",
      "GCA_002335545.1_Aspe_assembly01 No NEAT stop\n",
      "GCA_002564285.1_Tym_barrerae_AO245_v1 No NEAT start\n",
      "GCA_002564285.1_Tym_barrerae_AO245_v1 No MALAT start\n",
      "GCA_002564285.1_Tym_barrerae_AO245_v1 No NEAT stop\n",
      "GCA_002564305.1_Oct_mimax_AO248_v1 No NEAT start\n",
      "GCA_002564305.1_Oct_mimax_AO248_v1 No MALAT start\n",
      "GCA_002564305.1_Oct_mimax_AO248_v1 No NEAT stop\n",
      "GCA_002564305.1_Oct_mimax_AO248_v1 No MALAT stop\n",
      "GCA_002901085.1_ASM290108v1 No MALAT start\n",
      "GCA_002995585.1_ASM299558v1 No NEAT start\n",
      "GCA_002995585.1_ASM299558v1 No NEAT stop\n",
      "GCA_003033725.1_ASM303372v1 No NEAT start\n",
      "GCA_003033725.1_ASM303372v1 No NEAT stop\n",
      "GCA_003118495.1_macFus_1.0 No NEAT start\n",
      "GCA_003118495.1_macFus_1.0 No NEAT stop\n",
      "GCA_003182575.1_ASM318257v1 No NEAT start\n",
      "GCA_003182575.1_ASM318257v1 No NEAT stop\n",
      "GCA_003254725.2_ASM325472v2 No MALAT start\n",
      "GCA_003697985.1_ASM369798v1 No NEAT start\n",
      "GCA_003697985.1_ASM369798v1 No NEAT stop\n",
      "GCA_003697985.1_ASM369798v1 No MALAT stop\n",
      "GCA_003697995.1_ASM369799v1 No NEAT start\n",
      "GCA_003697995.1_ASM369799v1 No NEAT stop\n",
      "GCA_004023965.1_SpiGra_v1_BIUU No NEAT start\n",
      "GCA_004023965.1_SpiGra_v1_BIUU No NEAT stop\n",
      "GCA_004024585.1_ParHer_v1_BIUU No NEAT start\n",
      "GCA_004024585.1_ParHer_v1_BIUU No NEAT stop\n",
      "GCA_004024625.1_MelCap_v1_BIUU No NEAT start\n",
      "GCA_004024625.1_MelCap_v1_BIUU No NEAT stop\n",
      "GCA_004024645.1_MizCoq_v1_BIUU No NEAT start\n",
      "GCA_004024645.1_MizCoq_v1_BIUU No NEAT stop\n",
      "GCA_004024685.1_DipSte_v1_BIUU No NEAT start\n",
      "GCA_004024685.1_DipSte_v1_BIUU No NEAT stop\n",
      "GCA_004024765.1_ZapHud_v1_BIUU No NEAT start\n",
      "GCA_004024765.1_ZapHud_v1_BIUU No MALAT start\n",
      "GCA_004024765.1_ZapHud_v1_BIUU No NEAT stop\n",
      "GCA_004024805.1_XerIna_v1_BIUU No NEAT start\n",
      "GCA_004024805.1_XerIna_v1_BIUU No NEAT stop\n",
      "GCA_004024825.1_PygNem_v1_BIUU No NEAT start\n",
      "GCA_004024825.1_PygNem_v1_BIUU No NEAT stop\n",
      "GCA_004024885.1_SagImp_v1_BIUU No NEAT start\n",
      "GCA_004024885.1_SagImp_v1_BIUU No NEAT stop\n",
      "GCA_004024905.1_TapInd_v1_BIUU No NEAT start\n",
      "GCA_004024905.1_TapInd_v1_BIUU No NEAT stop\n",
      "GCA_004024925.1_ScaAqu_v1_BIUU No NEAT start\n",
      "GCA_004024925.1_ScaAqu_v1_BIUU No NEAT stop\n",
      "GCA_004024945.1_UroGra_v1_BIUU No NEAT start\n",
      "GCA_004024945.1_UroGra_v1_BIUU No MALAT start\n",
      "GCA_004024945.1_UroGra_v1_BIUU No NEAT stop\n",
      "GCA_004024985.1_SaiTat_v1_BIUU No NEAT start\n",
      "GCA_004024985.1_SaiTat_v1_BIUU No NEAT stop\n",
      "GCA_004024985.1_SaiTat_v1_BIUU No MALAT stop\n",
      "GCA_004025005.1_TadBra_v1_BIUU No NEAT start\n",
      "GCA_004025005.1_TadBra_v1_BIUU No MALAT start\n",
      "GCA_004025005.1_TadBra_v1_BIUU No NEAT stop\n",
      "GCA_004025065.1_SemEnt_v1_BIUU No NEAT start\n",
      "GCA_004025065.1_SemEnt_v1_BIUU No NEAT stop\n",
      "GCA_004025085.1_ThrSwi_v1_BIUU No NEAT start\n",
      "GCA_004025085.1_ThrSwi_v1_BIUU No NEAT stop\n",
      "GCA_004026545.1_MorMeg_v1_BIUU No NEAT start\n",
      "GCA_004026545.1_MorMeg_v1_BIUU No NEAT stop\n",
      "GCA_004026565.1_RanTarSib_v1_BIUU No NEAT start\n",
      "GCA_004026565.1_RanTarSib_v1_BIUU No NEAT stop\n",
      "GCA_004026665.1_MurFea_v1_BIUU No NEAT start\n",
      "GCA_004026665.1_MurFea_v1_BIUU No NEAT stop\n",
      "GCA_004026705.1_MicTal_v1_BIUU No NEAT start\n",
      "GCA_004026705.1_MicTal_v1_BIUU No MALAT start\n",
      "GCA_004026705.1_MicTal_v1_BIUU No NEAT stop\n",
      "GCA_004026745.1_MyrTri_v1_BIUU No NEAT start\n",
      "GCA_004026745.1_MyrTri_v1_BIUU No MALAT start\n",
      "GCA_004026745.1_MyrTri_v1_BIUU No NEAT stop\n",
      "GCA_004026765.1_MicHir_v1_BIUU No NEAT start\n",
      "GCA_004026765.1_MicHir_v1_BIUU No NEAT stop\n",
      "GCA_004026805.1_LasBor_v1_BIUU No NEAT start\n",
      "GCA_004026805.1_LasBor_v1_BIUU No MALAT start\n",
      "GCA_004026805.1_LasBor_v1_BIUU No NEAT stop\n",
      "GCA_004026805.1_LasBor_v1_BIUU No MALAT stop\n",
      "GCA_004026825.1_HemHyl_v1_BIUU No NEAT start\n",
      "GCA_004026825.1_HemHyl_v1_BIUU No NEAT stop\n",
      "GCA_004026855.1_LepAme_v1_BIUU No NEAT start\n",
      "GCA_004026855.1_LepAme_v1_BIUU No MALAT start\n",
      "GCA_004026855.1_LepAme_v1_BIUU No NEAT stop\n",
      "GCA_004026885.1_MegLyr_v1_BIUU No NEAT start\n",
      "GCA_004026885.1_MegLyr_v1_BIUU No NEAT stop\n",
      "GCA_004026905.1_HysCri_v1_BIUU No NEAT start\n",
      "GCA_004026905.1_HysCri_v1_BIUU No MALAT start\n",
      "GCA_004026905.1_HysCri_v1_BIUU No NEAT stop\n",
      "GCA_004026945.1_OviCan_v1_BIUU No NEAT start\n",
      "GCA_004026945.1_OviCan_v1_BIUU No NEAT stop\n",
      "GCA_004026965.1_PetTyp_v1_BIUU No NEAT start\n",
      "GCA_004026965.1_PetTyp_v1_BIUU No NEAT stop\n",
      "GCA_004027005.1_MusAve_v1_BIUU No NEAT start\n",
      "GCA_004027005.1_MusAve_v1_BIUU No NEAT stop\n",
      "GCA_004027005.1_MusAve_v1_BIUU No MALAT stop\n",
      "GCA_004027025.1_MyoCoy_v1_BIUU No NEAT start\n",
      "GCA_004027025.1_MyoCoy_v1_BIUU No MALAT start\n",
      "GCA_004027025.1_MyoCoy_v1_BIUU No NEAT stop\n",
      "GCA_004027085.1_MesBid_v1_BIUU No NEAT start\n",
      "GCA_004027165.1_CteSoc_v1_BIUU No NEAT start\n",
      "GCA_004027165.1_CteSoc_v1_BIUU No NEAT stop\n",
      "GCA_004027185.1_GliGli_v1_BIUU No NEAT start\n",
      "GCA_004027185.1_GliGli_v1_BIUU No MALAT start\n",
      "GCA_004027185.1_GliGli_v1_BIUU No NEAT stop\n",
      "GCA_004027185.1_GliGli_v1_BIUU No MALAT stop\n",
      "GCA_004027275.1_EulFul_v1_BIUU No NEAT start\n",
      "GCA_004027275.1_EulFul_v1_BIUU No NEAT stop\n",
      "GCA_004027295.1_DolPat_v1_BIUU No NEAT start\n",
      "GCA_004027295.1_DolPat_v1_BIUU No MALAT start\n",
      "GCA_004027295.1_DolPat_v1_BIUU No NEAT stop\n",
      "GCA_004027315.2_DicBicMic_v2_BIUU_UCD No MALAT start\n",
      "GCA_004027375.1_MacSob_v1_BIUU No NEAT start\n",
      "GCA_004027375.1_MacSob_v1_BIUU No NEAT stop\n",
      "GCA_004027395.1_CanFam_VD_v1_BIUU No NEAT start\n",
      "GCA_004027395.1_CanFam_VD_v1_BIUU No MALAT start\n",
      "GCA_004027395.1_CanFam_VD_v1_BIUU No NEAT stop\n",
      "GCA_004027455.1_HydHyd_v1_BIUU No NEAT start\n",
      "GCA_004027455.1_HydHyd_v1_BIUU No MALAT start\n",
      "GCA_004027455.1_HydHyd_v1_BIUU No NEAT stop\n",
      "GCA_004027495.1_BeaHun_v1_BIUU No NEAT start\n",
      "GCA_004027495.1_BeaHun_v1_BIUU No NEAT stop\n",
      "GCA_004027555.1_CraTho_v1_BIUU No NEAT start\n",
      "GCA_004027555.1_CraTho_v1_BIUU No NEAT stop\n",
      "GCA_004027595.1_DinBra_v1_BIUU No MALAT start\n",
      "GCA_004027615.1_CertNeg_v1_BIUU No NEAT start\n",
      "GCA_004027635.1_CroInd_v1_BIUU No NEAT start\n",
      "GCA_004027635.1_CroInd_v1_BIUU No MALAT start\n",
      "GCA_004027635.1_CroInd_v1_BIUU No NEAT stop\n",
      "GCA_004027655.1_GraMur_v1_BIUU No NEAT start\n",
      "GCA_004027655.1_GraMur_v1_BIUU No MALAT start\n",
      "GCA_004027655.1_GraMur_v1_BIUU No NEAT stop\n",
      "GCA_004027655.1_GraMur_v1_BIUU No MALAT stop\n",
      "GCA_004027695.1_CavTsc_v1_BIUU No NEAT start\n",
      "GCA_004027695.1_CavTsc_v1_BIUU No MALAT start\n",
      "GCA_004027695.1_CavTsc_v1_BIUU No NEAT stop\n",
      "GCA_004027715.1_CalDon_v1_BIUU No NEAT start\n",
      "GCA_004027715.1_CalDon_v1_BIUU No NEAT stop\n",
      "GCA_004027735.1_CarPer_v1_BIUU No NEAT start\n",
      "GCA_004027735.1_CarPer_v1_BIUU No MALAT start\n",
      "GCA_004027735.1_CarPer_v1_BIUU No NEAT stop\n",
      "GCA_004027775.1_BraVar_v1_BIUU No NEAT start\n",
      "GCA_004027775.1_BraVar_v1_BIUU No MALAT start\n",
      "GCA_004027775.1_BraVar_v1_BIUU No NEAT stop\n",
      "GCA_004027775.1_BraVar_v1_BIUU No MALAT stop\n",
      "GCA_004027835.1_AloPal_v1_BIUU No NEAT start\n",
      "GCA_004027835.1_AloPal_v1_BIUU No NEAT stop\n",
      "GCA_004027895.1_AllBul_v1_BIUU No NEAT start\n",
      "GCA_004027895.1_AllBul_v1_BIUU No NEAT stop\n",
      "GCA_004027915.1_CapPil_v1_BIUU No NEAT start\n",
      "GCA_004027915.1_CapPil_v1_BIUU No MALAT start\n",
      "GCA_004027915.1_CapPil_v1_BIUU No NEAT stop\n",
      "GCA_004027955.1_ChaVel_v1_BIUU No NEAT start\n",
      "GCA_004027955.1_ChaVel_v1_BIUU No NEAT stop\n",
      "GCA_004216515.1_sis1-161031-pseudohap No MALAT start\n",
      "GCA_004363435.1_PlaMin_v1_BIUU No NEAT start\n",
      "GCA_004363435.1_PlaMin_v1_BIUU No NEAT stop\n",
      "GCA_004363515.1_IniGeo_v1_BIUU No NEAT start\n",
      "GCA_004363515.1_IniGeo_v1_BIUU No NEAT stop\n",
      "GCA_004363535.1_DasPun_v1_BIUU No NEAT start\n",
      "GCA_004363535.1_DasPun_v1_BIUU No MALAT start\n",
      "GCA_004363535.1_DasPun_v1_BIUU No NEAT stop\n",
      "GCA_004363575.1_SolPar_v1_BIUU No MALAT start\n",
      "GCA_004363605.1_IndInd_v1_BIUU No NEAT start\n",
      "GCA_004363605.1_IndInd_v1_BIUU No MALAT start\n",
      "GCA_004364475.1_ZipCav_v1_BIUU No NEAT start\n",
      "GCA_004364475.1_ZipCav_v1_BIUU No NEAT stop\n",
      "GCA_004365215.1_CunPac_v1_BIUU No NEAT start\n",
      "GCA_004365215.1_CunPac_v1_BIUU No MALAT start\n",
      "GCA_004365215.1_CunPac_v1_BIUU No NEAT stop\n",
      "GCA_004886185.2_Basenji_breed-1.1 No MALAT start\n",
      "GCA_005280165.1_ASM528016v1 No MALAT stop\n",
      "GCA_005406085.2_Prionailurus_bengalensis_euptilurus_v03 No NEAT start\n",
      "GCA_005406085.2_Prionailurus_bengalensis_euptilurus_v03 No NEAT stop\n",
      "GCA_005458795.1_GSC_VIM_1.0 No NEAT start\n",
      "GCA_005458795.1_GSC_VIM_1.0 No MALAT start\n",
      "GCA_006408545.1_HBT No NEAT start\n",
      "GCA_006408545.1_HBT No MALAT start\n",
      "GCA_006408545.1_HBT No NEAT stop\n",
      "GCA_006408545.1_HBT No MALAT stop\n",
      "GCA_006408615.1_BWD No NEAT start\n",
      "GCA_006408615.1_BWD No NEAT stop\n",
      "GCA_006408675.1_KDD No NEAT start\n",
      "GCA_006408675.1_KDD No NEAT stop\n",
      "GCA_006408735.1_CMD No MALAT start\n",
      "GCA_006408755.1_TMG No NEAT start\n",
      "GCA_006408755.1_TMG No NEAT stop\n",
      "GCA_006410495.1_BUB No NEAT start\n",
      "GCA_006410495.1_BUB No MALAT start\n",
      "GCA_006410495.1_BUB No NEAT stop\n",
      "GCA_006410495.1_BUB No MALAT stop\n",
      "GCA_006410555.1_IBX No NEAT start\n",
      "GCA_006410555.1_IBX No MALAT start\n",
      "GCA_006410555.1_IBX No NEAT stop\n",
      "GCA_006410555.1_IBX No MALAT stop\n",
      "GCA_006410615.1_SUN No NEAT start\n",
      "GCA_006410615.1_SUN No MALAT start\n",
      "GCA_006410615.1_SUN No NEAT stop\n",
      "GCA_006410615.1_SUN No MALAT stop\n",
      "GCA_006410655.1_DFW No NEAT start\n",
      "GCA_006410655.1_DFW No NEAT stop\n",
      "GCA_006410675.1_KSP No NEAT start\n",
      "GCA_006410675.1_KSP No MALAT start\n",
      "GCA_006410675.1_KSP No NEAT stop\n",
      "GCA_006410695.1_MXD No NEAT start\n",
      "GCA_006410695.1_MXD No MALAT stop\n",
      "GCA_006410775.1_LEK No NEAT start\n",
      "GCA_006410775.1_LEK No MALAT start\n",
      "GCA_006410775.1_LEK No NEAT stop\n",
      "GCA_006410775.1_LEK No MALAT stop\n",
      "GCA_006410875.1_RAL No NEAT start\n",
      "GCA_006410875.1_RAL No MALAT start\n",
      "GCA_006410875.1_RAL No NEAT stop\n",
      "GCA_006410935.1_BHR No NEAT start\n",
      "GCA_006410935.1_BHR No MALAT start\n",
      "GCA_006410935.1_BHR No NEAT stop\n",
      "GCA_006410935.1_BHR No MALAT stop\n",
      "GCA_006411685.1_MTN No NEAT start\n",
      "GCA_006411685.1_MTN No NEAT stop\n",
      "GCA_006411685.1_MTN No MALAT stop\n",
      "GCA_006416875.1_CME No NEAT start\n",
      "GCA_006416875.1_CME No MALAT start\n",
      "GCA_006416875.1_CME No NEAT stop\n",
      "GCA_006417275.1_ORB No NEAT start\n",
      "GCA_006417275.1_ORB No MALAT start\n",
      "GCA_006417275.1_ORB No NEAT stop\n",
      "GCA_006417275.1_ORB No MALAT stop\n",
      "GCA_006461725.1_NPU_MCHR_1.0 No NEAT start\n",
      "GCA_006461725.1_NPU_MCHR_1.0 No NEAT stop\n",
      "GCA_007455615.1_ASM745561v1 No NEAT start\n",
      "GCA_007455615.1_ASM745561v1 No NEAT stop\n",
      "GCA_007922755.1_PedCap_v1_BIUU No NEAT start\n",
      "GCA_007922755.1_PedCap_v1_BIUU No NEAT stop\n",
      "GCA_007922795.1_NycHum_v1_BIUU No NEAT start\n",
      "GCA_007922795.1_NycHum_v1_BIUU No MALAT start\n",
      "GCA_007922795.1_NycHum_v1_BIUU No NEAT stop\n",
      "GCA_007922815.1_MacCal_v1_BIUU No MALAT start\n",
      "GCA_008641055.3_ASM864105v3 No MALAT start\n",
      "GCA_008750895.1_Mzaz_1.0 No NEAT start\n",
      "GCA_008750895.1_Mzaz_1.0 No NEAT stop\n",
      "GCA_008750915.1_Mjon_1.0 No NEAT start\n",
      "GCA_008750915.1_Mjon_1.0 No NEAT stop\n",
      "GCA_008750935.1_Mtav_1.0 No NEAT start\n",
      "GCA_008750935.1_Mtav_1.0 No MALAT start\n",
      "GCA_008750935.1_Mtav_1.0 No NEAT stop\n",
      "GCA_008750955.1_Mmit_1.0 No NEAT start\n",
      "GCA_008750955.1_Mmit_1.0 No MALAT start\n",
      "GCA_008750955.1_Mmit_1.0 No NEAT stop\n",
      "GCA_008750975.1_Mrav_1.0 No NEAT start\n",
      "GCA_008750975.1_Mrav_1.0 No MALAT start\n",
      "GCA_008750975.1_Mrav_1.0 No NEAT stop\n",
      "GCA_008750995.1_Mgri_1.0 No NEAT start\n",
      "GCA_008750995.1_Mgri_1.0 No MALAT start\n",
      "GCA_008750995.1_Mgri_1.0 No NEAT stop\n",
      "GCA_009758055.1_ASM975805v1 No MALAT stop\n",
      "GCA_009823505.1_RhiPru_1.0 No MALAT start\n",
      "GCA_009823505.1_RhiPru_1.0 No MALAT stop\n",
      "GCA_010120015.1_ASM1012001v1 No MALAT start\n",
      "GCA_011327655.1_ASM1132765v1 No NEAT start\n",
      "GCA_011327655.1_ASM1132765v1 No MALAT start\n",
      "GCA_011327655.1_ASM1132765v1 No NEAT stop\n",
      "GCA_011634725.1_ASM1163472v1 No NEAT start\n",
      "GCA_011634725.1_ASM1163472v1 No MALAT start\n",
      "GCA_011634725.1_ASM1163472v1 No NEAT stop\n",
      "GCA_011680675.1_LBP_v1 No MALAT start\n",
      "GCA_011751065.1_L.cinereus_Cryan_1219_p1.0 No MALAT start\n",
      "GCA_011751065.1_L.cinereus_Cryan_1219_p1.0 No MALAT stop\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "GCA_011754075.1_ASM1175407v1 No NEAT start\n",
      "GCA_011754075.1_ASM1175407v1 No NEAT stop\n",
      "GCA_012044875.1_ASM1204487v1 No MALAT start\n",
      "GCA_012045015.1_ASM1204501v1 No MALAT start\n",
      "GCA_013276365.2_UNSW_CanFamBas_1.2 No MALAT start\n",
      "GCA_014363405.1_P_pselaphon_scaffold_01 No MALAT start\n",
      "GCA_014364545.1_U_thibetanus_scaffold_01 No NEAT start\n",
      "GCA_014364545.1_U_thibetanus_scaffold_01 No NEAT stop\n",
      "GCA_014441545.1_ROS_Cfam_1.0 No MALAT start\n",
      "GCA_015711505.1_Sylvilagus_bachmani_HiC No NEAT start\n",
      "GCA_015711505.1_Sylvilagus_bachmani_HiC No MALAT start\n",
      "GCA_016801295.1_DSBC_Mcra_1.0 No NEAT start\n",
      "GCA_016801295.1_DSBC_Mcra_1.0 No NEAT stop\n",
      "GCA_016801295.1_DSBC_Mcra_1.0 No MALAT stop\n",
      "GCA_017311385.1_HRRL_Platanista_gangetica_1.0 No NEAT start\n",
      "GCA_017311385.1_HRRL_Platanista_gangetica_1.0 No MALAT start\n",
      "GCA_017311385.1_HRRL_Platanista_gangetica_1.0 No NEAT stop\n",
      "GCA_017311455.1_Otocyon_megalotis_TS305_17_09_2019 No NEAT start\n",
      "GCA_017311455.1_Otocyon_megalotis_TS305_17_09_2019 No MALAT start\n",
      "GCA_017311455.1_Otocyon_megalotis_TS305_17_09_2019 No NEAT stop\n",
      "GCA_018403415.1_P_iriomotensis_scaffold_01 No NEAT start\n",
      "GCA_019843815.1_ASM1984381v1 No NEAT start\n",
      "GCA_019843815.1_ASM1984381v1 No NEAT stop\n",
      "GCA_019843835.1_ASM1984383v1 No NEAT start\n",
      "GCA_019843835.1_ASM1984383v1 No MALAT start\n",
      "GCA_019843835.1_ASM1984383v1 No NEAT stop\n",
      "GCA_019843835.1_ASM1984383v1 No MALAT stop\n",
      "GCA_019843855.1_ASM1984385v1 No NEAT start\n",
      "GCA_019843855.1_ASM1984385v1 No MALAT start\n",
      "GCA_019843855.1_ASM1984385v1 No NEAT stop\n",
      "GCA_019843855.1_ASM1984385v1 No MALAT stop\n",
      "GCA_019844195.1_ASM1984419v1 No MALAT stop\n",
      "GCA_019903745.2_ULRtarCaribou_2v2 No NEAT start\n",
      "GCA_019903745.2_ULRtarCaribou_2v2 No NEAT stop\n",
      "GCA_019903745.2_ULRtarCaribou_2v2 No MALAT stop\n",
      "GCA_020387435.1_ASM2038743v1 No NEAT start\n",
      "GCA_020392405.1_ASM2039240v1 No MALAT start\n",
      "GCA_020740815.1_VCOM_Snig_1.0 No NEAT start\n",
      "GCA_020740815.1_VCOM_Snig_1.0 No NEAT stop\n",
      "GCA_020854095.1_UniMelb_DasViv_v1.0 No MALAT start\n",
      "GCA_021234165.1_CSHL_Meso_final No MALAT start\n",
      "GCA_021461705.1_Cryptotis_parva_assembly_1.0 No NEAT start\n",
      "GCA_021461705.1_Cryptotis_parva_assembly_1.0 No NEAT stop\n",
      "GCA_021461705.1_Cryptotis_parva_assembly_1.0 No MALAT stop\n",
      "GCA_023159225.1_ASM2315922v1 No NEAT start\n",
      "GCA_023170115.1_Speothos_venaticus_v01 No NEAT start\n",
      "GCA_023170115.1_Speothos_venaticus_v01 No MALAT start\n",
      "GCA_023170115.1_Speothos_venaticus_v01 No NEAT stop\n",
      "GCA_023548195.1_mBetpen1.pri.20210916 No MALAT start\n",
      "GCA_023553655.1_mMyrfas1.20211206 No MALAT start\n",
      "GCA_023553655.1_mMyrfas1.20211206 No NEAT stop\n",
      "GCA_023764695.1_ASM2376469v1 No NEAT stop\n",
      "GCA_023783135.1_ASM2378313v1 No MALAT start\n",
      "GCA_023783435.1_ASM2378343v1 No MALAT start\n",
      "GCA_023783475.1_ASM2378347v1 No MALAT start\n",
      "GCA_023898255.1_ASM2389825v1 No MALAT start\n",
      "GCA_024139225.1_mSunEtr1.pri.cur No NEAT start\n",
      "GCA_024363435.1_mPerLon1.0.a No NEAT start\n",
      "GCA_024363435.1_mPerLon1.0.a No MALAT start\n",
      "GCA_024363575.1_mPerLon1.0.p No NEAT start\n",
      "GCA_024363575.1_mPerLon1.0.p No MALAT start\n",
      "GCA_024711535.1_mDipMer1.0.p No NEAT start\n",
      "GCA_024803745.1_mThoBot1.0.p No NEAT start\n",
      "GCA_024803745.1_mThoBot1.0.p No MALAT start\n",
      "GCA_025584935.1_MNHNYannick_LA_1 No NEAT start\n",
      "GCA_025584935.1_MNHNYannick_LA_1 No NEAT stop\n",
      "GCA_026826555.1_mTolMat1 No MALAT start\n",
      "GCA_027406575.1_mNycCou1.pri No MALAT start\n",
      "GCA_027563665.1_DD_ASM_mAP_20210503 No MALAT start\n",
      "GCA_028533205.1_Lagorchestes_hirsutus_HiC No MALAT start\n",
      "GCA_028533215.1_Babyrousa_celebensis_HiC No NEAT start\n",
      "GCA_028533215.1_Babyrousa_celebensis_HiC No NEAT stop\n",
      "GCA_028533335.1_Chrysocyon_brachyurus_HiC No MALAT start\n",
      "GCA_028533375.1_Diceros_bicornis_HiC No NEAT start\n",
      "GCA_028533375.1_Diceros_bicornis_HiC No MALAT start\n",
      "GCA_028533765.1_Pteropus_rufus_HiC No MALAT start\n",
      "GCA_028533885.1_Nasua_narica_HiC No NEAT stop\n",
      "GCA_028533885.1_Nasua_narica_HiC No MALAT stop\n",
      "GCA_028534055.1_Eulemur_mongoz_HiC No MALAT start\n",
      "GCA_028565675.1_ASM2856567v1 No MALAT start\n",
      "GCA_028583105.1_mf-2k No NEAT start\n",
      "GCA_028583685.1_PetGlider_PUasm1.0 No MALAT start\n",
      "GCA_028627135.1_Pseudochirops_cupreus_HiC No MALAT start\n",
      "GCA_028627135.1_Pseudochirops_cupreus_HiC No NEAT stop\n",
      "GCA_028627215.1_mg-2k No NEAT start\n",
      "GCA_028627265.1_Cercopithecus_mitis_HiC No NEAT start\n",
      "GCA_028627265.1_Cercopithecus_mitis_HiC No NEAT stop\n",
      "GCA_028644305.1_Xerus_rutilus_HiC No NEAT start\n",
      "GCA_028646385.1_Steno_bredanensis_HiC No NEAT stop\n",
      "GCA_028646425.1_Grampus_griseus_HiC No NEAT start\n",
      "GCA_028646515.1_Pseudochirops_corinnae_HiC No MALAT start\n",
      "GCA_028646535.1_pl-1k No NEAT stop\n",
      "GCA_028646575.1_Pseudocheirus_occidentalis_HiC No NEAT start\n",
      "GCA_028646595.1_pg-2k No MALAT start\n",
      "GCA_028658305.1_Pachyuromys_duprasi_HiC No MALAT start\n",
      "GCA_028658325.1_Potorous_gilbertii_HiC No MALAT start\n",
      "GCA_029224305.1_Cmar_C18_SBIK-F_TBG_1 No NEAT start\n",
      "GCA_029224305.1_Cmar_C18_SBIK-F_TBG_1 No NEAT stop\n",
      "GCA_029224305.1_Cmar_C18_SBIK-F_TBG_1 No MALAT stop\n",
      "GCA_029532125.1_Ltowsendii_DMNS18807_Prim_1.0 No NEAT start\n",
      "GCA_029532125.1_Ltowsendii_DMNS18807_Prim_1.0 No MALAT start\n",
      "GCA_029593785.1_MVZ_CabUni_1.0 No NEAT start\n",
      "GCA_029593785.1_MVZ_CabUni_1.0 No MALAT start\n",
      "GCA_029593785.1_MVZ_CabUni_1.0 No NEAT stop\n",
      "GCA_029958975.1_ASM2995897v1 No MALAT start\n",
      "GCA_030035585.1_mDugDug1.hap1 No MALAT start\n",
      "GCA_900006375.2_Gulo_2.2_annotated No NEAT start\n",
      "GCA_900006375.2_Gulo_2.2_annotated No NEAT stop\n",
      "GCA_900661375.1_LYPA1.0 No NEAT stop\n",
      "GCA_902168325.1_Pnud_10x_v1 No NEAT start\n",
      "GCA_902168325.1_Pnud_10x_v1 No MALAT start\n",
      "GCA_902168325.1_Pnud_10x_v1 No NEAT stop\n",
      "GCA_902168325.1_Pnud_10x_v1 No MALAT stop\n",
      "GCA_902168405.1_Pazt_10x_v1 No NEAT start\n",
      "GCA_902168405.1_Pazt_10x_v1 No MALAT start\n",
      "GCA_902168405.1_Pazt_10x_v1 No NEAT stop\n",
      "GCA_902168405.1_Pazt_10x_v1 No MALAT stop\n",
      "GCA_902168415.1_Pmel_10x_v1 No NEAT start\n",
      "GCA_902168415.1_Pmel_10x_v1 No MALAT start\n",
      "GCA_902168415.1_Pmel_10x_v1 No NEAT stop\n",
      "GCA_902168415.1_Pmel_10x_v1 No MALAT stop\n",
      "GCA_902168425.1_Patt_10x_v1 No NEAT start\n",
      "GCA_902168425.1_Patt_10x_v1 No MALAT start\n",
      "GCA_902168425.1_Patt_10x_v1 No NEAT stop\n",
      "GCA_902168425.1_Patt_10x_v1 No MALAT stop\n",
      "GCA_902729485.2_Mus_minutoides_I2396_v1 No NEAT start\n",
      "GCA_902729485.2_Mus_minutoides_I2396_v1 No NEAT stop\n",
      "GCA_903992545.1_mPipPip1.1 No NEAT start\n",
      "GCA_903992545.1_mPipPip1.1 No MALAT start\n",
      "GCA_903992545.1_mPipPip1.1 No NEAT stop\n",
      "GCA_903992545.1_mPipPip1.1 No MALAT stop\n",
      "GCA_905146905.1_NYPRO_anot_genome No MALAT start\n",
      "GCA_907164565.1_mPsaObe1.curated_primary_1811 No MALAT start\n",
      "GCA_949987585.1_mPipPyg2.1 No NEAT start\n",
      "GCF_000001905.1_Loxafr3.0 No NEAT start\n",
      "GCF_000002285.5_Dog10K_Boxer_Tasha No MALAT start\n",
      "GCF_000002295.2_MonDom5 No MALAT start\n",
      "GCF_000002295.2_MonDom5 No NEAT stop\n",
      "GCF_000002295.2_MonDom5 No MALAT stop\n",
      "GCF_000003625.3_OryCun2.0 No NEAT start\n",
      "GCF_000003625.3_OryCun2.0 No MALAT start\n",
      "GCF_000147115.1_Myoluc2.0 No NEAT start\n",
      "GCF_000151735.1_Cavpor3.0 No MALAT start\n",
      "GCF_000151845.1_Pvam_2.0 No MALAT start\n",
      "GCF_000164805.1_Tarsius_syrichta-2.0.1 No NEAT start\n",
      "GCF_000164805.1_Tarsius_syrichta-2.0.1 No NEAT stop\n",
      "GCF_000165445.2_Mmur_3.0 No MALAT start\n",
      "GCF_000208655.1_Dasnov3.0 No MALAT start\n",
      "GCF_000223135.1_CriGri_1.0 No MALAT stop\n",
      "GCF_000260355.1_ConCri1.0 No NEAT start\n",
      "GCF_000296755.1_EriEur2.0 No MALAT start\n",
      "GCF_000299155.1_EleEdw1.0 No MALAT start\n",
      "GCF_000313985.2_ASM31398v2 No MALAT start\n",
      "GCF_000325575.1_ASM32557v1 No MALAT start\n",
      "GCF_000327345.1_ASM32734v1 No NEAT start\n",
      "GCF_000412655.1_ASM41265v1 No NEAT start\n",
      "GCF_000622305.1_S.galili_v1.0 No MALAT start\n",
      "GCF_000696695.1_Burgud No NEAT start\n",
      "GCF_000754665.1_Bison_UMD1.0 No NEAT start\n",
      "GCF_000754665.1_Bison_UMD1.0 No NEAT stop\n",
      "GCF_001458135.1_marMar2.1 No NEAT start\n",
      "GCF_002099425.1_phaCin_unsw_v4.1 No MALAT start\n",
      "GCF_002204375.1_MunDraft-v1.0 No MALAT start\n",
      "GCF_003255815.1_Tgel_1.0 No NEAT stop\n",
      "GCF_006229205.1_meerkat_22Aug2017_6uvM2_HiC No MALAT start\n",
      "GCF_012274545.1_DMR_v1.0_HiC No MALAT start\n",
      "GCF_012295265.1_UNSW_AlpineDingo_1.0 No MALAT start\n",
      "GCF_014108245.1_mPipKuh1.p No NEAT start\n",
      "GCF_014441545.1_ROS_Cfam_1.0 No MALAT start\n",
      "GCF_014570535.1_YNU_ManJav_2.0 No MALAT stop\n",
      "GCF_014570555.1_YNU_ManPten_2.0 No MALAT stop\n",
      "GCF_014633375.1_OchPri4.0 No MALAT start\n",
      "GCF_014824575.2_WHU_Shon_v2.1 No MALAT start\n",
      "GCF_015220235.1_mChoDid1.pri No MALAT start\n",
      "GCF_015852505.1_mTacAcu1.pri No MALAT start\n",
      "GCF_016433145.1_AgileGrace No MALAT start\n",
      "GCF_017591425.1_NIBS_Ocur_1.0 No MALAT start\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "GCF_018350155.1_O.geoffroyi_Oge1_pat1.0 No MALAT stop\n",
      "GCF_019393635.1_mDroGli1.pri No MALAT start\n",
      "GCF_900094665.1_CAROLI_EIJ_v1.1 No MALAT stop\n",
      "GCF_902635505.1_mSarHar1.11 No MALAT start\n",
      "GCF_902729225.1_Ma_sr-lr_union100 No MALAT start\n"
     ]
    }
   ],
   "source": [
    "NEAT = {}\n",
    "MALAT = {}\n",
    "# headers of sequences in 'known_NEAT1_MALAT1_and_features.fasta' grouped by types\n",
    "# group of known NEAT1 orthologs\n",
    "a = ['Human_NEAT1','Marsupials_NEAT1','Mouse_NEAT1']\n",
    "# group of known MALAT1 orthologs\n",
    "b = ['Human_MALAT1','Mouse_MALAT1']\n",
    "# group of NEAT1 tRNA-like structures\n",
    "c = ['AfterTriplex_Human_NEAT1','AfterTriplex_Marsupials_NEAT1','AfterTriplex_Mouse_NEAT1']\n",
    "# group of MALAT1 promotor areas\n",
    "d = ['MALAT_start_human','MALAT_start_mouse','MALAT_Tata']\n",
    "# group of NEAT1 promotor areas\n",
    "d1 = ['Marsupials_NEAT1_start','NEAT_Tata']\n",
    "\n",
    "# Parsing blastn outputs to identify approximate coordinates of orthologs\n",
    "for lib in libs:\n",
    "    with open('./BLAST_features/' + lib + '.txt') as file1:\n",
    "        N = {}\n",
    "        M = {}\n",
    "        N_start = {}\n",
    "        M_start = {}\n",
    "        N_stop = {}\n",
    "        M_stop = {}\n",
    "        NTata = {}\n",
    "        MTata = {}\n",
    "        NTrip = {}\n",
    "        MTrip = {}\n",
    "        \n",
    "        for line in file1:\n",
    "            dat = line.strip().split('\\t')\n",
    "            if dat[0] in a:\n",
    "                if dat[1] not in N:\n",
    "                    N[dat[1]] = 0\n",
    "                N[dat[1]] += int(dat[2])\n",
    "                if dat[4] == '1':\n",
    "                    N_start[dat[1]] = dat[6]\n",
    "            if dat[0] in b:\n",
    "                if dat[1] not in M:\n",
    "                    M[dat[1]] = 0\n",
    "                M[dat[1]] += int(dat[2])\n",
    "                if dat[4] == '1':\n",
    "                    M_start[dat[1]] = dat[6]\n",
    "            if dat[0] in d1 and int(dat[2]) > 30 and dat[1] not in NTata:\n",
    "                NTata[dat[1]] = dat[6]\n",
    "            if dat[0] in d and int(dat[2]) > 30 and dat[1] not in MTata:\n",
    "                MTata[dat[1]] = dat[6]\n",
    "            if dat[0] == 'AfterTriplex_Human_MALAT1' and dat[4] == '1' and dat[1] not in M_stop:\n",
    "                M_stop[dat[1]] = dat[6]\n",
    "            if dat[0] in c and dat[4] == '1' and dat[1] not in N_stop:\n",
    "                N_stop[dat[1]] = dat[6]\n",
    "            if dat[0] == 'NEAT_triplex' and dat[5] == '52' and dat[1] not in NTrip:\n",
    "                NTrip[dat[1]] = dat[7]\n",
    "            if dat[0] == 'MALAT_triplex' and dat[5] == '92' and dat[1] not in MTrip:\n",
    "                MTrip[dat[1]] = dat[7]\n",
    "                \n",
    "        # Summarise the results of parsing\n",
    "        Ncont = sorted(N.keys(), key=lambda kv: N[kv],reverse=True)[0]\n",
    "        Mcont = sorted(M.keys(), key=lambda kv: M[kv],reverse=True)[0]\n",
    "        if Ncont == Mcont:\n",
    "            NEAT[lib] = {'contig': Ncont}\n",
    "            MALAT[lib] = {'contig': Mcont}\n",
    "        else:\n",
    "            Ncont = Mcont\n",
    "            NEAT[lib] = {'contig': 'FromMALAT_' + Ncont}\n",
    "            MALAT[lib] = {'contig': Mcont}\n",
    "            \n",
    "        if Ncont in NTata:\n",
    "            NEAT[lib]['start'] = NTata[Ncont]\n",
    "        else:\n",
    "            if Ncont in N_start:\n",
    "                NEAT[lib]['start'] = N_start[Ncont]\n",
    "            else:\n",
    "                print(lib,'No NEAT start')\n",
    "        if Ncont in MTata:\n",
    "            MALAT[lib]['start'] = MTata[Ncont]\n",
    "        else:\n",
    "            if Ncont in M_start:\n",
    "                MALAT[lib]['start'] = M_start[Ncont]\n",
    "            else:\n",
    "                print(lib,'No MALAT start')\n",
    "        if Ncont in N_stop:\n",
    "            NEAT[lib]['stop'] = N_stop[Ncont]\n",
    "        else:\n",
    "            if Ncont in NTrip:\n",
    "                NEAT[lib]['stop'] = NTrip[Ncont]\n",
    "            else:\n",
    "                print(lib,'No NEAT stop')\n",
    "        if Ncont in M_stop:\n",
    "            MALAT[lib]['stop'] = M_stop[Ncont]\n",
    "        else:\n",
    "            if Ncont in MTrip:\n",
    "                MALAT[lib]['stop'] = MTrip[Ncont]\n",
    "            else:\n",
    "                print(lib,'No MALAT stop')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "b5f0ba73",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "83 171\n"
     ]
    }
   ],
   "source": [
    "# collect genomes without start or stop coordinates\n",
    "left = []\n",
    "for lib in NEAT:\n",
    "    if not 'start' in NEAT[lib] or not 'stop' in NEAT[lib]:\n",
    "        left.append(lib)\n",
    "\n",
    "# Check for long (> 5kb) homology patches between known orthologs\n",
    "# and genomic assemblies lacking stop or start (or both) coordinates\n",
    "\n",
    "manual = []\n",
    "for lib in left:\n",
    "    with open('./BLAST_features/' + lib + '.txt') as file1:\n",
    "        for line in file1:\n",
    "            dat = line.strip().split('\\t')\n",
    "            if int(dat[2]) > 5000:\n",
    "                manual.append(lib)\n",
    "                break\n",
    "                \n",
    "print(len(manual), len(left))\n",
    "\n",
    "#keep for the analysis only assemblies with clear homology\n",
    "to_exclude = [i for i in left if i not in manual]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "433e807e",
   "metadata": {},
   "outputs": [],
   "source": [
    "#write into file the results of coordinates search\n",
    "\n",
    "#GenomeID Neat1ContigAccession Neat1Start Neat1Stop Neat1ApproxLength Malat1ContigAccession Malat1Start\n",
    "#Malat1Stop Malat1ApproxLength\n",
    "\n",
    "#GCA_000283155.1_CerSimSim1.0\tJH767791.1\t9929556\t9952059\t22503\tJH767791.1\t9997904\t10005062\t7158\n",
    "\n",
    "out = open('coordinates_algorithmic','w')\n",
    "for lib in libs:\n",
    "    if lib not in to_exclude:\n",
    "        l = [lib]\n",
    "        if lib in NEAT:\n",
    "            l.append(NEAT[lib]['contig'])\n",
    "            if 'start' in NEAT[lib]:\n",
    "                l.append(NEAT[lib]['start'])\n",
    "            else:\n",
    "                l.append('Nan')\n",
    "\n",
    "            if 'stop' in NEAT[lib]:\n",
    "                l.append(NEAT[lib]['stop'])\n",
    "            else:\n",
    "                l.append('Nan')\n",
    "            if 'start' in NEAT[lib] and 'stop' in NEAT[lib]:\n",
    "                l.append(str(abs(int(NEAT[lib]['start']) - int(NEAT[lib]['stop']))))\n",
    "            else:\n",
    "                l.append('Nan')\n",
    "        else:\n",
    "            l += ['Nan','Nan','Nan']\n",
    "        if lib in MALAT:\n",
    "            l.append(MALAT[lib]['contig'])\n",
    "            if 'start' in MALAT[lib]:\n",
    "                l.append(MALAT[lib]['start'])\n",
    "            else:\n",
    "                l.append('Nan')\n",
    "\n",
    "            if 'stop' in MALAT[lib]:\n",
    "                l.append(MALAT[lib]['stop'])\n",
    "            else:\n",
    "                l.append('Nan')\n",
    "            if 'start' in MALAT[lib] and 'stop' in MALAT[lib]:\n",
    "                l.append(str(abs(int(MALAT[lib]['start']) - int(MALAT[lib]['stop']))))\n",
    "            else:\n",
    "                l.append('Nan')\n",
    "        else:\n",
    "            l += ['Nan','Nan','Nan']\n",
    "        out.write('\\t'.join(l) + '\\n')\n",
    "\n",
    "out.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "27d0cedb",
   "metadata": {},
   "source": [
    "### Manual curation of approximate coordinates\n",
    "After that we performed a manual curation of the cases where start or stop coordinate was missing by looking at the results of blastn search. In some cases the contig assembly was incomplete and a gene was lacking 5' or 3' end, if the majority of the gene was preserved we set the missing coordinate to 0 and kept the ortholog in the dataset. In other cases we could clearly see many patches of homology to known orthologs and only the promotor area did not give the hit in blastn search; if we could approximate the location of the ortholog, we set the start coordinate manually. "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "aa6c000b",
   "metadata": {},
   "source": [
    "### Assigning taxonomic tree to the species with identified orthologs after manual curation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "7475f529",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "544\n"
     ]
    }
   ],
   "source": [
    "# ManualCuration file is the 'coordinates_algorithmic' after manually added start/stop coordinates\n",
    "sel = {}\n",
    "with open('ManualCuration') as file1:\n",
    "    for line in file1:\n",
    "        if not line.startswith('#'):\n",
    "            dat = line.strip().split('\\t')\n",
    "            sel[dat[0]] = {}\n",
    "            sel[dat[0]]['coor'] = dat[1:]\n",
    "print(len(sel))            "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "183a660f",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Names of known NEAT1 and MALAT1 orthologs\n",
    "ref_names = ['NEAT1_Mouse','NEAT1_Marsupials','NEAT1_Human','MALAT1_Mouse','MALAT1_Human']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "1bd700d3",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Retrieve taxid for mammals with identified orthologs\n",
    "with open('Namesystem') as file1:\n",
    "    for line in file1:\n",
    "        dat = line.strip().split('\\t')\n",
    "        if dat[0] in sel:\n",
    "            sel[dat[0]]['sc_name'] = dat[1]\n",
    "            sel[dat[0]]['taxid'] = dat[2]\n",
    "        if dat[0] in ref_names:\n",
    "            sel[dat[0]] = {}\n",
    "            sel[dat[0]]['sc_name'] = dat[1]\n",
    "            sel[dat[0]]['taxid'] = dat[2] \n",
    "taxid_list = [sel[i]['taxid'] for i in sel]  "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "97d76f71",
   "metadata": {},
   "source": [
    "Below files (*.dpm) are part of the taxonomy package from NCBI\n",
    "https://ftp.ncbi.nlm.nih.gov/pub/taxonomy/taxdump.tar.gz"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "0bf3f4ae",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Function to loop through the taxonomy tree\n",
    "def parse_nodes(taxid, data_dict):\n",
    "    if taxid in parent_dict and taxid in rank_dict:\n",
    "        while parent_dict[taxid] != taxid:\n",
    "            data_dict[rank_dict[taxid]] = names_dict[taxid]\n",
    "            taxid = parent_dict[taxid]\n",
    "    else:\n",
    "          data_dict = {}  \n",
    "    return(data_dict)\n",
    "categories = [\"species\", \"genus\", \"family\", \"order\", \"class\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "cc81f9b3",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Collect common and scientific names from NCBI files and hierarchy of taxa\n",
    "comname = {}\n",
    "rank_dict = {}\n",
    "parent_dict = {}\n",
    "names_dict = {}\n",
    "with open('/home/ksenia/databases/Taxonomy/names.dmp') as names_file:\n",
    "    names = csv.reader(names_file,delimiter='\\t',quoting=csv.QUOTE_NONE)     \n",
    "    for line in names:\n",
    "        if line[6] == 'scientific name':\n",
    "            names_dict[line[0]] = line[2]\n",
    "        if 'common name' in line[6]:\n",
    "            if line[0] in taxid_list:\n",
    "                comname[line[0]] = line[2]\n",
    "\n",
    "with open('/home/ksenia/databases/Taxonomy/nodes.dmp','r')as nodes_file:\n",
    "    nodes = csv.reader(nodes_file,delimiter='\\t',quoting=csv.QUOTE_NONE)\n",
    "    for line in nodes:\n",
    "        parent_dict[line[0]] = line[2]\n",
    "        rank_dict[line[0]] = line[4]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "id": "8fba2832",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Build a taxonomy tree for individual genome of our collection\n",
    "taxonomy = {}\n",
    "for item in taxid_list:\n",
    "    taxonomy[item] = {}\n",
    "    z = parse_nodes(item, {})\n",
    "    if z:\n",
    "        for cat in categories:\n",
    "            if cat in z.keys():\n",
    "                taxonomy[item][cat] = z[cat] \n",
    "            else:\n",
    "                taxonomy[item][cat] = 'NA'\n",
    "    else:\n",
    "        for cat in categories:\n",
    "            taxonomy[item][cat] = 'NA'\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "id": "8ff14786",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Create the first metadata file\n",
    "out = open('./Key_Files/Metadata','w')\n",
    "out.write('#File_name\\tScientific_name\\tCommon_name\\ttaxid\\t' + '\\t'.join(categories[::-1]) + \n",
    "          'NEAT1Contig\\tNEAT1start\\tNEAT1stop\\t' + \n",
    "          'MALAT1Contig\\tMALAT1start\\tMALAT1stop\\n')\n",
    "for item in sel:\n",
    "    l = [item, sel[item]['sc_name']]\n",
    "    if sel[item]['taxid'] in comname:\n",
    "        l.append(comname[sel[item]['taxid']])\n",
    "    else:\n",
    "        l.append('Nan')\n",
    "    l.append(sel[item]['taxid'])\n",
    "    for cat in categories[::-1]:\n",
    "        l.append(taxonomy[sel[item]['taxid']][cat])\n",
    "    l += sel[item]['coor']\n",
    "    out.write('\\t'.join(l) + '\\n')\n",
    "    \n",
    "out.close()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
